%\VignetteIndexEntry{Gviz users guide}
%\VignetteDepends{Gviz,EnsDb.Hsapiens.v75}
%\VignetteKeywords{visualization, genomics}
%\VignettePackage{Gviz}
\documentclass[11pt]{article}

% A bunch of styles and package requirements for the Bioconductor vignette branding
<<style, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex()
@

\RequirePackage[utf8]{inputenc}
\SweaveOpts{keep.source=FALSE,pdf=TRUE,eps=FALSE}
% \RequirePackage{hyperref}
\RequirePackage{url}
%\RequirePackage[authoryear,round]{natbib}
%\bibliographystyle{plainnat}
% \RequirePackage[text={7.2in,9in},centering]{geometry}
\RequirePackage{Sweave}
\setkeys{Gin}{width=0.95\textwidth}
\RequirePackage{longtable}
\RequirePackage{graphicx}
\newcommand{\code}[1]{{\texttt{#1}}}
\newcommand{\term}[1]{{\emph{#1}}}
\newcommand{\Rmethod}[1]{{\textit{#1}}}
\newcommand{\Rfunarg}[1]{{\textit{#1}}}
\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}
\newcommand{\mgg}[0]{\Rpackage{Gviz} }
\newcommand{\Reference}[1]{{\texttt{#1}}}
\newcommand{\link}[1]{{#1}}
\newcommand{\RR}[0]{{\texttt{R}}}

\title{The Gviz User Guide}
\author{Florian Hahne\footnote{florian.hahne@novartis.com}}
\date{Edited: November 2015; Compiled: \today}

\begin{document}

<<init, echo=FALSE, results=hide>>=
options(width=65)
library(xtable)
source(system.file("scripts/documentation.R", package="Gviz"))
xtabDetails <- details
addParTable <- function(class, skip=c("showTitle", "size", "background.title"), add=NULL)
{
    Parameters <- data.frame("Display Parameter"=names(xtabDetails[[class]]),
                             "Description"=xtabDetails[[class]], check.names=FALSE)
    align <- "lrp{5in}"
    if(!is.null(add)){
        Parameters <- cbind(Parameters, add)
        align <- c("lp{4in}", "lp{4in}", "lp{4in}", "lp{4in}")
    }
    Parameters <- Parameters[order(Parameters[,1]),]
    Parameters <- apply(Parameters, 2, function(x) gsub("_", "\\_", x, fixed=TRUE))
    rownames(Parameters) <-  gsub("_", "\\_", rownames(Parameters), fixed=TRUE)
    sel <- Parameters[,1] %in% skip
    Parameters[,2] <- gsub("\\code{\\linkS4class{", "\\Rclass{{", Parameters[,2], fixed=TRUE)
    print(xtable(Parameters[!sel,], align=align), sanitize.text.function=function(x) x, include.rownames=FALSE,
          floating=FALSE, tabular.environment="longtable")
    return(invisible())
}

hasUcscConnection <- !is(try(rtracklayer::browserSession(), silent=TRUE), "try-error")
oto <- options(timeout=5)
## Use the ensembl mirror of biomart since biomart seems to be offline.
hasBiomartConnection <- (!is(try(download.file("http://www.biomart.org", tempfile(), quiet=TRUE)), "try-error") &&
                          !is(try(biomaRt::listMarts(host="www.ensembl.org"), silent=TRUE),
                              "try-error"))
options(timeout=oto)
## Uncommenting this helps when the UCSC server has a hickup but still lets you connect:
##hasUcscConnection <- !is(try(rtracklayer::browserSession(), silent=TRUE), "try-error") && !is(try(IdeogramTrack(genome="hg19", chromosome=7), silent=TRUE), "try-error")

@

\maketitle

\tableofcontents
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}

In order to make sense of genomic data one often aims to plot such
data in a genome browser, along with a variety of genomic annotation
features, such as gene or transcript models, CpG island, repeat
regions, and so on. These features may either be extracted from public
data bases like ENSEMBL or UCSC, or they may be generated or curated
in-house. Many of the currently available genome browsers do a
reasonable job in displaying genome annotation data, and there are
options to connect to some of them from within \R (e.g., using the
\Rpackage{rtracklayer} package). However, none of these solutions
offer the flexibility of the full \RR graphics system to display large
numeric data in a multitude of different ways. The \mgg package aims
to close this gap by providing a structured visualization framework to
plot any type of data along genomic coordinates. It is loosely based
on the \Rpackage{GenomeGraphs} package by Steffen Durinck and James
Bullard, however the complete class hierarchy as well as all the
plotting methods have been restructured in order to increase
performance and flexibility. All plotting is done using the grid
graphics system, and several specialized annotation classes allow to
integrate publicly available genomic annotation data from sources like
UCSC or ENSEMBL.


\section{Basic Features}

The fundamental concept behind the \mgg package is similar to the
approach taken by most genome browsers, in that individual types of
genomic features or data are represented by separate tracks. Within
the package, each track constitutes a single object inheriting from
class \Rclass{GdObject}, and there are constructor functions as well
as a broad range of methods to interact with and to plot these
tracks. When combining multiple objects, the individual tracks will
always share the same genomic coordinate system, thus taking the
burden of aligning the plot elements from the user.

It is worth mentioning that, at a given time, tracks in the sense of the
\mgg package are only defined for a single chromosome on a specific
genome, at least for the duration of a given plotting operation. You will later see
that a track may still contain information for multiple chromosomes,
however most of this is hidden except for the currently active
chromosome, and the user will have to explicitely switch the chromsome
to access the inactive parts. While the package in principle imposes
no fixed structure on the chromosome or on the genome names, it makes
sense to stick to a standaradized naming paradigm, in particular when
fetching additional annotation information from online resources. By
default this is enforced by a global option
\code{ucscChromosomeNames}, which is set during package loading and
which causes the package to check all supplied chromosome names for
validity in the sense of the UCSC definition (chromosomes have to
start with the \term{chr} string). You may decide to turn this feature
off by calling \code{options(ucscChromosomeNames=FALSE)}. For the
remainder of this vignette however, we will make use of the UCSC
genome and chromosome identifiers, e.g., the \Robject{chr7} chromosome
on the mouse \Robject{mm9} genome.

The different track classes will be described in more detail in the
\Reference{Track classes} section further below. For now, let's just
take a look at a typical \mgg session to get an idea of what this is
all about. We begin our presentation of the available functionality by
loading the package:

<<loadPackage>>=
library(Gviz)
@

The most simple genomic features consist of start and stop
coordinates, possibly overlapping each other. CpG islands or
microarray probes are real life examples for this class of
features. In the Bioconductor world those are most often represented
as run-length encoded vectors, for instance in the \Rclass{IRanges}
and \Rclass{GRanges} classes. To seamlessly integrate with other
Bioconductor packages, we can use the same data structures to generate
our track objects. A sample set of CpG island coordinates has been
saved in the \Robject{cpgIslands} object and we can use that for our
first annotation track object. The constructor function
\Rfunction{AnnotationTrack} is a convenient helper to create the
object.

<<AnnotationTrack>>=
library(GenomicRanges)
data(cpgIslands)
class(cpgIslands)
chr <- as.character(unique(seqnames(cpgIslands)))
gen <- genome(cpgIslands)
atrack <- AnnotationTrack(cpgIslands, name="CpG")
@

Please note that the \Rfunction{AnnotationTrack} constructor (as most
constructors in this package) is fairly flexible and can accomodate
many different types of inputs. For instance, the start and end
coordinates of the annotation features could be passed in as
individual arguments \Robject{start} and \Robject{end}, as a
\Rclass{data.frame} or even as an \Rclass{IRanges} or
\Rclass{GRangesList} object. Furthermore, a whole bunch of coercion
methods are available for those package users that prefer the more
traditional R coding paradigm, and they should allow operations along
the lines of \code{as(obj, 'AnnotationTrack')}. You may want to
consult the class' manual page for more information, or take a look at
the \Reference{Bioconductor integration} section for a listing of the
most common data structures and their respective counterparts in the
\mgg package.

With our first track object being created we may now proceed to the
plotting. There is a single function \Rfunction{plotTracks} that
handles all of this. As we will learn in the remainder of this
vignette, \Rfunction{plotTracks} is quite powerful and has a number of
very useful additional arguments. For now we will keep things very
simple and just plot the single CpG islands annotation track.

<<plotAnnotationTrack, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(atrack)
@

As you can see, the resulting graph is not particularly
spectacular. There is a title region showing the track's name on a
gray background on the left side of the plot and a data region showing
the seven individual CpG islands on the right. This structure is
similar for all the available track objects classes and it somewhat
mimicks the layout of the popular UCSC Genome Browser. If you are not
happy with the default settings, the \mgg
package offers a multitude of options to fine-tune the track
appearance, which will be shown in the \Reference{Plotting Parameters}
section.

Appart from the relative distance of the Cpg islands, this
visualization does not tell us much. One obvious next step would be to
indicate the genomic coordinates we are currently looking at in order
to provide some reference. For this purpose, the \mgg package offers
the \Rclass{GenomeAxisTrack} class. Objects from the class can be
created using the constructor function of the same name.

<<GenomeAxisTrack>>=
gtrack <- GenomeAxisTrack()
@

Since a \Rclass{GenomeAxisTrack} object is always relative to the
other tracks that are plotted, there is little need for additional
arguments. Essentially, the object just tells the
\Rfunction{plotTracks} function to add a genomic axis to the
plot. Nonetheless, it represent a separate annotation track just as
the CpG island track does. We can pass this additional track on to
\Rfunction{plotTracks} in the form of a list.

<<plotGenomeAxisTrack, fig=TRUE, width=7.5, height=1.1>>=
plotTracks(list(gtrack, atrack))
@

You may have realized that the genomic axis does not take up half of
the available vertical plotting space, but only uses the space
necessary to fit the axis and labels. Also the title region for this
track is empty. In general, the \mgg package tries to find reasonable
defaults for all the parameters controlling the look and feel of a
plots so that appealing visualizations can be created without too much
tinkering. However, all features on the plot including the relative
track sizes can also be adjusted manually.

As mentioned before in the beginning of this vignette, a plotted track
is always defined for exactly one chromosome on a particular
genome. We can include this information in our plot by means of a
chromosome ideogram. An ideogram is a simplified visual representation
of a chromosome, with the different chromosomal staining bands
indicated by color, and the centromer (if present) indicated by the
shape. The necessary information to produce this visualization is
stored in online data repositories, for instance at UCSC. The \mgg
package offers very convenient connections to some of these
repositories, and the \Rclass{IdeogramTrack} constructor function is
one example for such a connection. With just the information about a
valid UCSC genome and chromosome, we can directly fetch the chromosome
ideogram information and construct a dedicated track object that can
be visualized by \Rfunction{plotTracks}. Please not that you will need
an established internet connection for this to work, and that fetching
data from UCSC can take quite a long time, depending on the server
load. The \mgg package tries to cache as much data as possible to
reduce the bandwidth in future queries.

<<showIdeogramTrack, eval=FALSE>>=
itrack <- IdeogramTrack(genome=gen, chromosome=chr)
@

<<doIdeogramTrack, echo=FALSE, results=hide>>=
if(hasUcscConnection){
    itrack <- IdeogramTrack(genome=gen, chromosome=chr)
}else{
    data(itrack)
}
@

Similar to the previous examples, we stick the additional track object
into a list in order to plot it.

<<plotIdeogramTrack, fig=TRUE, width=7.5, height=1.5>>=
plotTracks(list(itrack, gtrack, atrack))
@

Ideogram tracks are the one exception in all of \mgg's track objects
in the sense that they are not really displayed on the same coordinate
system like all the other tracks. Instead, the current genomic
location is indicated on the chromosome by a red box (or, as in this
case, a red line if the width is too small to fit a box).

So far we have only looked at very basic annotation features and how
to give a point of reference to our plots. Naturally, we also want to
be able to handle more complex genomic features, such as gene
models. One potential use case would be to utilize gene model
information from an existing local source. Alternatively, we could
dowload such data from one of the available online resources like UCSC
or ENSEBML, and there are constructor functions to handle these
tasks. For this example we are going to load gene model data from a
stored \Rclass{data.frame}. The track class of choice here is a
\Rclass{GeneRegionTrack} object, which can be created via the
constructor function of the same name. Similar to the
\Rclass{AnnotationTrack} constructor there are multiple possible ways
to pass in the data.

<<GeneRegionTrack, fig=TRUE, width=7.5, height=3>>=
data(geneModels)
grtrack <- GeneRegionTrack(geneModels, genome=gen, chromosome=chr, name="Gene Model")
plotTracks(list(itrack, gtrack, atrack, grtrack))
@

In all those previous examples the plotted genomic range has been
determined automatically from the input tracks. Unless told otherwise,
the package will always display the region from the leftmost item to
the rightmost item in any of the tracks.  Of course such a static view
on a chromosomal region is of rather limited use. We often want to
zoom in or out on a particular plotting region to see more details or
to get a broader overview. To that end, \Rfunction{plotTracks}
supports the \Robject{from} and \Robject{to} arguments that let us
choose an arbitrary genomic range to plot.

<<zooming, fig=TRUE, width=7.5, height=2.2>>=
plotTracks(list(itrack, gtrack, atrack, grtrack), from=26700000, to=26750000)
@

Another pair of arguments that controls the zoom state are
\Robject{extend.left} and \Robject{extend.right}. Rather than
\Robject{from} and \Robject{to}, those arguments are relative to the
currently displayed ranges, and can be used to quickly extend the view
on one or both ends of the plot. In addition to positive or negative
absolute integer values one can also provide a float value between -1
and 1 which will be interpreted as a zoom factor, i.e., a value of 0.5
will cause zooming in to half the currently displayed range.

<<zooming2, fig=TRUE, width=7.5, height=3>>=
plotTracks(list(itrack, gtrack, atrack, grtrack), extend.left=0.5, extend.right=1000000)
@

You may have noticed that the layout of the gene model track has
changed depending on the zoom level. This is a feature of the \mgg
package, which automatically tries to find the optimal visualization
settings to make best use of the available space. At the same time,
when features on a track are too close together to be plotted as
separate items with the current device resolution, the package will
try to reasonably merge them in order to avoid overplotting.

Often individual ranges on a plot tend to grow quite narrow, in
particular when zooming far out, and a couple of tweaks become helpful
in order to get nice plots, for instance to drop the bounding borders
of the exons.

<<zooming3, fig=TRUE, width=7.5, height=3>>=
plotTracks(list(itrack, gtrack, atrack, grtrack), extend.left=0.5, extend.right=1000000, col=NULL)
@

When zooming further in it may become interesting to take a look at
the actual genomic sequence at a given position, and the \mgg package
provides the track class \Rclass{SequenceTrack} that let's you do just
that. Among several other options it can draw the necessary sequence
information from one of the \Rclass{BSgenome} packages.

<<zooming4, fig=TRUE, width=7.5, height=3.1>>=
library(BSgenome.Hsapiens.UCSC.hg19)
strack <- SequenceTrack(Hsapiens, chromosome=chr)
plotTracks(list(itrack, gtrack, atrack, grtrack, strack), from=26591822, to=26591852, cex=0.8)
@

So far we have replicated the features of a whole bunch of other
genome browser tools out there. The real power of the package comes
with a rather general track type, the
\Rclass{DataTrack}. \Rclass{DataTrack} object are essentially
run-length encoded numeric vectors or matrices, and we can use them to
add all sorts of numeric data to our genomic coordinate plots. There
are a whole bunch of different visualization options for these tracks,
from dot plots to histograms to box-and-whisker plots. The individual
rows in a numeric matrix are considered to be different data groups or
samples, and the columns are the raster intervals in the genomic
coordinates. Of course, the data points (or rather the data ranges) do
not have to be evenly spaced; each column is associated with a
particular genomic location. For demonstration purposes we can create
a simple \Rclass{DataTrack} object from randomly sampled data.

<<DataTrack, fig=TRUE, width=7.5, height=4>>=
set.seed(255)
lim <- c(26700000, 26750000)
coords <- sort(c(lim[1], sample(seq(from=lim[1], to=lim[2]), 99), lim[2]))
dat <- runif(100, min=-10, max=10)
dtrack <- DataTrack(data=dat, start=coords[-length(coords)], end=coords[-1], chromosome=chr,
                    genome=gen, name="Uniform")
plotTracks(list(itrack, gtrack, atrack, grtrack, dtrack), from=lim[1], to=lim[2])
@

The first thing to notice is that the title panel to the right of the
plot now contains a y-axis indicating the range of the displayed
data. The default plotting type for numeric vectors is a simple dot
plot. This is by far not the only visualization option, and in a sense
it is waisting quite a lot of information because the run-length
encoded ranges are not immediately apparent. We can change the plot
type by supplying the \Robject{type} argument to
\Rfunction{plotTracks}. A complete description of the available
plotting options is given in section \Reference{Track classes}, and a
more detailed treatment of the so-called 'display parameters' that
control the look and feel of a track is given in the
\Reference{Plotting Parameters} section.

<<DataTrackHist, fig=TRUE, width=7.5, height=4>>=
plotTracks(list(itrack, gtrack, atrack, grtrack, dtrack), from=lim[1], to=lim[2], type="histogram")
@

As we can see, the data values in the numeric vector are indeed
matched to the genomic coordinates of the \Rclass{DataTrack}
object. Such a visualization can be particularly helpful when
displaying for instance the coverage of NGS reads along a chromosome,
or to show the measurement values of mapped probes from a micro array
experiment.

This concludes our first introduction into the \mgg
package. The remainder of this vignette will deal in much more depth
with the different concepts and the various track classes and plotting
options.

\section{Plotting parameters}

\subsection{Setting parameters}

Although not implicitely mentioned before, we have already made use of
the plotting parameter facilities in the \mgg package, or, as we will
call them from now on, the 'display parameters'. Display parameters
are properties of individual track objects (i.e., of any object
inheriting from the base \Rclass{GdObject} class). They can either be
set during object instantiation as additional arguments to the
constructor functions or, for existing track objects, using the
\Rfunction{displayPars} replacement method. In the former case, all
named arguments that can not be matched to any of the constructor's
formal arguments are considered to be display parameters, regardless of
their type or whether they are defined for a particular track class or
not. The following code example rebuilds our \Rclass{GeneRegionTrack}
object with a bunch of display parameters and demonstrates the use of
the \Rfunction{displayPars} accessor and replacement methods.

<<displayPars1f, fig=TRUE, width=7.5, height=3>>=
grtrack <- GeneRegionTrack(geneModels, genome=gen, chromosome=chr, name="Gene Model",
                           transcriptAnnotation="symbol", background.title="brown")
head(displayPars(grtrack))
displayPars(grtrack) <- list(background.panel="#FFFEDB", col=NULL)
head(displayPars(grtrack))

plotTracks(list(itrack, gtrack, atrack, grtrack))
@

For our gene model track we have now added the gene symbols of the
different transcripts to the plot, removed the gray border around the
individual exons models, and changed the background color of both the
title and the data panel to a yellowish hue. There is a third option
to set display parameters for a single plotting operation (rather than
the permanent setting in the track object) by passing in additional
arguments to the \Rfunction{plotTracks} function. We have already made
use of this feature, for instance in the previous data plotting type
example. It is worth mentioning that all display parameters which are
passed along with the \Rfunction{plotTracks} function apply to
\emph{all} track objects in the plot. For some objects classes a
particular display parameter may not make any sense, and in that case
it is simply ignored. Also, the settings only apply for one single
plotting operation and will not be retained in the plotted track
objects. They do however get precedence over the object-internal
parameters. The following line of code exemplifies this behaviour.

<<displayPars2f, fig=TRUE, width=7.5, height=3>>=
plotTracks(list(itrack, gtrack, atrack, grtrack),
           background.panel="#FFFEDB", background.title="darkblue")
@

In order to make full use of the flexible parameter system we need to
know which display parameters control which aspect of which track
class. The obvious source for this information are the man pages of
the respective track classes, which list all available parameters
along with a short description of their effect and default values in
the \Robject{Display Parameters} section. Alternatively, we can use
the \Rfunction{availableDisplayPars} function, which prints out the
available parameters for a class as well as their default values in a
list-like structure. The single argument to the function is either a
class name of a track object class, or the object itself, in which
case its class is automatically detected.

<<displayPars3>>=
dp <- availableDisplayPars(grtrack)
tail(dp)
@

As we can see, display parameters can be inherited from parent
classes. For the regular user this is not important at all, however it
nicely exemplifies the structure of the class hierarchy in the \mgg
package.

\subsection{Schemes}

Users might find themselves changing the same parameters over and over
again, and it would make sense to register these modifications in a
central location once and for all. To this end the \mgg package
supports display parameter schemes. A scheme is essentially just a
bunch of nested named lists, where the names on the first level of
nesting should correspond to track class names, and the names on the
second level to the display parameters to set. The currently active
schmeme can be changed by setting the global option
\Robject{Gviz.scheme}, and a new scheme can be registered by using the
\Rfunction{addScheme} function, providing both the list and the name
for the new scheme. The \Rfunction{getScheme} function is useful to
get the current scheme as a list structure, for instance to use as a
skeleton for your own custom scheme.

<<displayPars4, fig=TRUE, width=7.5, height=1.5>>=
getOption("Gviz.scheme")
scheme <- getScheme()
scheme$GeneRegionTrack$fill <- "salmon"
scheme$GeneRegionTrack$col <- NULL
scheme$GeneRegionTrack$transcriptAnnotation <- "transcript"
addScheme(scheme, "myScheme")
options(Gviz.scheme="myScheme")
grtrack <- GeneRegionTrack(geneModels, genome=gen, chromosome=chr, name="Gene Model")
plotTracks(grtrack)
options(Gviz.scheme="default")
grtrack <- GeneRegionTrack(geneModels, genome=gen, chromosome=chr, name="Gene Model",
                           transcriptAnnotation="symbol")
@

In order to make these settings persitant across R sessions one can
create one or several schemes in the global environment in the special
object \Robject{.GvizSchemes}, for instance by putting the necessary
code in the \Robject{.Rprofile} file. This object needs to be a named
list of schemes, and it will be collected when the \mgg package
loads. Its content is then automatically added to the collection of
available schemes.  The following pseudo-code exemplifies this and
could go into an \Robject{.Rprofile} file.

<<schemes, eval=FALSE>>=
.GvizSchemes <- list(myScheme=list(GeneRegionTrack=list(fill="salmon", col=NULL, transcriptAnnotation="transcript")))
@

Please note that because display parameters are stored
with the track objects, a scheme change only has an effect on those
objects that are created after the change has taken place.

\subsection{Plotting direction}

By default all tracks will be plotted in a 5' -> 3' direction. It
sometimes can be useful to actually show the data relative to the
opposite strand. To this end one can use the \Rfunarg{reverseStrand}
display parameter, which does just what its name suggests. Since the
combination of forward and reverse stranded tracks on a single plot
does not make too much sense, one should usually set this as a global
display parameter in the \Rfunction{plotTracks} function. The function
will however cast a warning if a mixture of forward and reverse strand
tracks has been passed in for plotting.

<<plottingdirections, fig=TRUE, width=7.5, height=3>>=
plotTracks(list(itrack, gtrack, atrack, grtrack), reverseStrand=TRUE)
@

As you can see, the fact that the data has been plotted on the reverse
strand is also reflected in the \Rclass{GenomeAxis} track.


\section{Track classes}

In this section we will highlight all of the available annotation
track classes in the \mgg package. For the complete reference of all
the nuts and bolts, including all the avaialable methods, please see
the respective class man pages. We will try to keep this vignette up
to date, but in cases of discrepancies between here and the man pages
you should assume the latter to be correct.

\subsection{GenomeAxisTrack}
\Rclass{GenomeAxisTrack} objects can be used to add some reference to
the currently displayed genomic location to a \mgg plot. In their most
basic form they are really just a horizontal axis with genomic
coordinate tickmarks. Using the \Rfunction{GenomeAxisTrack}
constructor function is the recommended way to instantiate objects
from the class. There is no need to know in advance about a particular
genomic location when constructing the object. Instead, the displayed
coordinates will be determined from the context, e.g., from the
\Rfunarg{from} and \Rfunarg{to} arguments of the
\Rfunction{plotTracks} function, or, when plotted together with other
track objects, from their genomic locations.

<<GenomeAxisTrackClass1, fig=TRUE, width=7.5, height=0.5>>=
axisTrack <- GenomeAxisTrack()
plotTracks(axisTrack, from=1e6, to=9e6)
@

As an optional feature one can highlight particular regions
on the axis, for instance to indicated stretches of N nucleotides or
gaps in genomic alignments. Such regions have to be supplied to the
optional \Rfunarg{range} argument of the constructor function as
either an \Rclass{GRanges} or an \Rclass{IRanges} object.

<<GenomeAxisTrackClass2, fig=TRUE, width=7.5, height=0.5>>=
axisTrack <- GenomeAxisTrack(range=IRanges(start=c(2e6, 4e6), end=c(3e6, 7e6),
                                           names=rep("N-stretch", 2)))
plotTracks(axisTrack, from=1e6, to=9e6)
@

If names have been supplied with the \Rfunarg{range} argument, those
can also be added to the plot.

<<GenomeAxisTrackClass2a, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(axisTrack, from=1e6, to=9e6, showId=TRUE)
@

\subsubsection*{Display parameters for GenomeAxisTrack objects}
There are a whole bunch of display parameters to alter the appearance
of \Rclass{GenomeAxisTrack} plots. For instance, one could add
directional indicators to the axis using the \Rfunarg{add53} and
\Rfunarg{add35} parameters.

<<GenomeAxisTrackClass3, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(axisTrack, from=1e6, to=9e6, add53=TRUE, add35=TRUE)
@

Sometimes the resolution of the tick marks is not sufficient, in which
case the \Rfunarg{littleTicks} argument can be used to have a more
fine-grained axis annotation.

<<GenomeAxisTrackClass4, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(axisTrack, from=1e6, to=9e6, add53=TRUE, add35=TRUE, littleTicks=TRUE)
@

The \mgg package tries to come up with reasonable defaults for the
axis annotation. In our previous example, the genomic coordinates are
indicated in megabases. We can control this via the \Rfunarg{exponent}
parameter, which takes an integer value greater then zero. The
location of the tick marks are displayed as a fraction of
$10^{exponent}$.

<<GenomeAxisTrackClass5, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(axisTrack, from=1e6, to=9e6, exponent=4)
@

Another useful parameter, \Rfunarg{labelPos} controls the arrangement
of the tick marks. It takes one of the values \code{alternating},
\code{revAlternating}, \code{above} or \code{below}. For instance we
could aline all tick marks underneath the axis.

<<GenomeAxisTrackClass6, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(axisTrack, from=1e6, to=9e6, labelPos="below")
@

Sometimes a full-blown axis is just too much, and all we really need
in the plot is a small scale indicator. We can change the appearance
of the \Rclass{GenomeAxisTrack} object to such a limited
representation by setting the \Rfunarg{scale} display
parameter. Typically, this will be a numeric value between 0 and 1,
which is interpreted as the fraction of the plotting region used for
the scale. The plotting method will apply some rounding to come up
with reasonable and human-readable values. For even more control we
can pass in a value larger than 1, which is considered to be an
absolute range length. In this case, the user is responsible for the
scale to actually fit in the current plotting range.

<<GenomeAxisTrackClass7, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(axisTrack, from=1e6, to=9e6, scale=0.5)
@

We still have control over the placement of the scale label via the
\Rfunarg{labelPos}, parameter, which now takes the values
\code{above}, \code{below} and \code{beside}.

<<GenomeAxisTrackClass8, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(axisTrack, from=1e6, to=9e6, scale=0.5, labelPos="below")
@

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{GenomeAxisTrack}
class by typing in \code{?GenomeAxisTrack} on the \RR command line.

<<GenomeAxisTrackClassTable, echo=FALSE, results=tex>>=
addParTable("GenomeAxisTrack")
@

\subsection{IdeogramTrack}
While a genomic axis provides helpful points of reference to a plot,
it is sometimes important to show the currently displayed region in
the broader context of the whole chromosme. Are we looking at distal
regions, or somewhere close to the centromer? And how much of the
complete chromosome is covered in our plot. To that end the \mgg
package defines the \Rclass{IdeogramTrack} class, which is an
idealized representation of a single chromosome. When plotted, these
track objects will always show the whole chromosome, regardless of the
selected genomic region. However, the displayed coordinates are
indicated by a box that sits on the ideogram image. The chromosomal
data necessary to draw the ideogram is not part of the \mgg package
itself, instead it is downloaded from an online source (UCSC). Thus it
is important to use both chromosome and genome names that are
recognizable in the UCSC data base when dealing with
\Rclass{IdeogramTrack} objects. You might want to consult the UCSC
webpage (\url{http://genome.ucsc.edu/}) or use the
\Rfunction{ucscGenomes} function in the \Rpackage{rtracklayer} package
for a listing of available genomes.

Assuming the chromosome data are available online, a simple call to
the \Rfunction{IdeogramTrack} constructor function including the
desired genome and chromosome name are enough to instantiate the
object. Since the connection to UCSC can be slow, the package tries to
cache data that has already been downloaded for the duration of the \R
session. If needed, the user can manually clear the cache by calling
the \Rfunction{clearSessionCache} function. Of course it is also
possible to construct \Rclass{IdeogramTrack} objects from local
data. Please see the class' man page for details.

<<IdeogramTrackClass1Show, eval=FALSE>>=
ideoTrack <- IdeogramTrack(genome="hg19", chromosome="chrX")
plotTracks(ideoTrack, from=85e6, to=129e6)
@

<<IdeogramTrackClass1Do, fig=TRUE, width=7.5, height=0.5, echo=FALSE, results=hide>>=
if(hasUcscConnection){
    ideoTrack <- IdeogramTrack(genome="hg19", chromosome="chrX")
}else{
    data(itrack)
}
plotTracks(ideoTrack, from=85e6, to=129e6)
@

We can turn off the explicit plotting of the chromosome name by
setting the \Rfunarg{showId} display parameter to \code{FALSE}.

<<IdeogramTrackClass2, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(ideoTrack, from=85e6, to=129e6, showId=FALSE)
@

The chromosome bands in the ideogram come with a unique identifier,
and we can add this information to the plot, at least for those bands
that are wide enought to accomodate the text.

<<IdeogramTrackClass3, fig=TRUE, width=7.5, height=0.5>>=
plotTracks(ideoTrack, from=85e6, to=129e6, showId=FALSE, showBandId=TRUE, cex.bands=0.5)
@



\subsubsection*{Display parameters for IdeogramTrack objects}

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{IdeogramTrack}
class by typing in \code{?IdeogramTrack} on the \RR command line.

<<IdeogramTrackClassTable, echo=FALSE, results=tex>>=
addParTable("IdeogramTrack")
@

\subsection{DataTrack}
Probably the most powerfull of all the track classes in the \mgg
package are \Rclass{DataTracks}. Essentially they constitute
run-length encoded numeric vectors or matrices, meaning that one or
several numeric values are associated to a particular genomic
coordinate range. These ranges may even be overlapping, for instance
when looking at results from a running window operation. There can be
multiple samples in a single data set, in which case the ranges are
associated to the columns of a numeric matrix rather than a numeric
vector, and the plotting method provides tools to incoorporate sample
group information. Thus the starting point for creating
\Rclass{DataTrack} objects will always be a set of ranges, either in
the form of an \Rclass{IRanges} or \Rclass{GRanges} object, or
individually as start and end coordinates or widths. The second
ingredient is a numeric vector of the same length as the number of
ranges, or a numeric matrix with the same number of columns. Those may
even already be part of the input \Rclass{GRanges} object as
\code{elemenMetadata} values. For a complete description of all the
possible inputs please see the class' online documentation. We can
pass all this information to the \Rfunction{DataTrack} constructor function to
instantiate an object. We will load our sample data from an
\Rclass{GRanges} object that comes as part of the \mgg package.

<<DataClass1, fig=TRUE, width=7.5, height=1.5>>=
data(twoGroups)
dTrack <- DataTrack(twoGroups, name="uniform")
plotTracks(dTrack)
@

The default visualization for our very simplistic sample
\Rclass{DataTrack} is a rather unispiring dot plot. The track comes
with a scale to indicate the range of the numeric values on the
y-axis, appart from that it looks very much like the previous
examples. A whole battery of display parameters is to our disposal to
control the track's look and feel. The most important one is the
\Rfunarg{type} parameter. It determines the type of plot to use and
takes one or several of the following values:

<<<types, echo=FALSE, results=tex>>=
types <- data.frame(Value=c("p", "l", "b", "a", "s", "S", "g", "r", "h", "confint", "smooth", "histogram", "mountain", "polygon", "boxplot", "gradient", "heatmap", "horizon"),
                    Type=c("dot plot", "lines plot", "dot and lines plot", "lines plot of average (i.e., mean) values", "stair steps (horizontal first)",
                           "stair steps (vertical first)", "add grid lines", "add linear regression line", "histogram lines", "confidence intervals for average values", "add loess curve",
                           "histogram (bar width equal to range with)", "'mountain-type' plot relative to a baseline",
                           "'polygon-type' plot relative to a baseline", "box and whisker plot",
                           "false color image of the summarized values", "false color image of the individual values",
                           "Horizon plot indicating magnitude and direction of a change relative to a baseline"))
print(xtable(types, align="lrp{5in}"), sanitize.text.function=function(x) x, include.rownames=FALSE,
          floating=FALSE, tabular.environment="longtable")
@

Displayed below are the same sample data as before but plotted with the different type settings:

<<typePlots, fig=TRUE, width=7.5, height=8.5, echo=FALSE, results=hide>>=
pushViewport(viewport(layout=grid.layout(nrow=9, ncol=2)))
i <- 1
for(t in types$Value)
{
    pushViewport(viewport(layout.pos.col=((i-1)%%2)+1, layout.pos.row=((i-1)%/%2)+1))
    if(t != "horizon"){
        names(dTrack) <- t
        plotTracks(dTrack, type=t, add=TRUE, cex.title=0.8, margin=0.5)
    }else{
        data(dtHoriz)
        names(dtHoriz) <- "horizon *"
        plotTracks(dtHoriz[8,], type="horizon", add=TRUE, cex.title=0.8, margin=0.5, showAxis=FALSE, horizon.origin=0.7)
    }
    i <- i+1
    popViewport(1)
}
popViewport(1)
names(dTrack) <- "uniform"
@

\renewcommand*{\thefootnote}{\fnsymbol{footnote}}
\footnotetext[1]{A different data set is plotted for the horizon type for the sake of clarity.}

You will notice that some of the plot types work better for univariate
data while others are clearly designed for multivariate inputs. The
\Rfunarg{a} type for instance averages the values at each genomic
location before plotting the derived values as a line. The decision
for a particular plot type is totally up to the user, and one could
even overlay multiple types by supplying a character vector rather
than a character scalar as the \Rfunarg(type) argument. For example,
this will combine a boxplot with an average line and a data grid.



<<mutitype, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack, type=c("boxplot", "a", "g"))
@

For the heatmap plotting type we arrange all the data in a
well-structured two-dimensional matrix which gives us the oportunity
to add a little extra information about the individual
samples. Depending on how the \Rclass{DataTrack} was created in the
first place we can choose to display the sample names (which in our
case correspond to the column names of the input \Rclass{GRanges}
object). The plot also highlights another feature of the heatmap type:
the y-axis now shows a mapping of the numeric values into the color
range.

<<sampNames, fig=TRUE, width=7.5, height=1.5>>=
colnames(mcols(twoGroups))
plotTracks(dTrack, type=c("heatmap"), showSampleNames=TRUE, cex.sampleNames=0.6)
@

\subsubsection*{Data Grouping}

An additional layer of flexibility is added by making use of \mgg's
grouping functionality. The individual samples (i.e., rows in the data
matrix) can be grouped together using a factor variable, and, if
reasonable, this grouping is reflected in the layout of the respective
track types. For instance our example data could be derived from two
different sample groups with three replicates each, and we could easily
integrate this information into our plot.

<<grouping, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack, groups=rep(c("control", "treated"), each=3), type=c("a", "p", "confint"))
@

For the dot plot representation the individual group levels are
indicated by color coding. For the \Rfunarg{a} type, the averages are
now computed for each group separately and also indicated by two lines
with similar color coding. Grouping is not supported for all plotting
types, for example the \Rfunarg{mountain} and \Rfunarg{polygon} type
already use color coding to convey a different message and for the
\Rfunarg{gradient} type the data are already collapsed to a single
variable. The following gives an overview over some of the other
groupable \Rclass{DataTrack} types. Please note that there are many
more display parameters that control the layout of both grouped and of
ungrouped \Rclass{DataTracks}. You may want to check the class' help
page for details.

<<typeGroupedPlots, fig=TRUE, width=7.5, height=6, echo=FALSE, results=hide>>=
pushViewport(viewport(layout=grid.layout(nrow=9, ncol=1)))
i <- 1
for(t in c("a", "s", "confint", "smooth", "histogram", "boxplot", "heatmap", "horizon"))
{
    pushViewport(viewport(layout.pos.col=((i-1)%%1)+1, layout.pos.row=((i-1)%/%1)+1))
    if(t != "horizon"){
        names(dTrack) <- t
        plotTracks(dTrack, type=t, add=TRUE, cex.title=0.8, groups=rep(1:2, each=3), margin=0.5)
    }else{
        plotTracks(dtHoriz[c(1,8),], type="horizon", add=TRUE, cex.title=0.8, margin=0.5, showAxis=FALSE, horizon.origin=0.3,
                   groups=1:2)
    }
    i <- i+1
    popViewport(1)
}
pushViewport(viewport(layout.pos.col=((i-1)%%1)+1, layout.pos.row=((i-1)%/%1)+1))
names(dTrack) <- "hor. hist."
plotTracks(dTrack, type="histogram", stackedBars=FALSE, add=TRUE, cex.title=0.8, groups=rep(1:2, each=3), margin=0.5)
popViewport(2)
names(dTrack) <- "uniform"
@

\footnotetext[1]{A different data set is plotted for the horizon type for the sake of clarity.}

If we need to display some additional information about the individual
group levels we can make use of the \Rfunarg{legend} display parameter
to add a simple legend to the plot. Depending on the plot type and on
some of the other display parameters, the look of this legend may vary
slightly.

<<groupingLegend, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack, groups=rep(c("control", "treated"), each=3), type=c("a", "p"), legend=TRUE)
@

For a grouped horizon plot the group labels have to be shown in a
similar fashion as for heatmaps, i.e., by setting the
\Rfunarg{showSampleNames} argument to \code{TRUE}.

<<horizLegend, fig=TRUE, results=hide, width=7.5, height=1.5>>=
data(dtHoriz)
dtHoriz <- dtHoriz[1:6,]
plotTracks(dtHoriz, type="horiz", groups=rownames(values(dtHoriz)),
           showSampleNames=TRUE, cex.sampleNames = 0.6, separator=1)
@

\subsubsection*{Building DataTrack objects from files}
A number of standard file types exist that all store numeric data
along genomic coordinates. We have tried to make such files accessible
in the \mgg package by providing additional options to the
\Rfunction{DataTrack} constructor function. In the previous examples
the \Rfunarg{range} argument was a \Rclass{GRanges} object. Instead,
we can also pass in the path to a file on disk by means of a character
scalar. The \Rclass{DataTrack} class supports the most common file
types like \code{wig}, \code{bigWig} or \code{bedGraph}, but also
knows how to deal with \code{bam} files. You may have realized that
some of these files are indexed, and we have taken the approach to
stream the data from indexed files on the fly when it is needed for
plotting.

However let's first start with the simple example of a \code{bedGraph}
file. These files support a single data sample, and thus are
equivalent to a \code{GRanges} object with a single numeric
metadata column. \code{bedGraph} files are not indexed,
so we have to load the whole file content when instantiating the
object.

<<filedt1, fig=TRUE, width=7.5, height=1>>=
bgFile <- system.file("extdata/test.bedGraph", package="Gviz")
dTrack2 <- DataTrack(range=bgFile, genome="hg19", type="l", chromosome="chr19", name="bedGraph")
class(dTrack2)
plotTracks(dTrack2)
@

As we can see the constructor has returned a regular
\Rclass{DataTrack} object. The function to be used in order to read
the data off the file has been automatically choosen by the package
based on the file extension of the input file. Of course the number of
these supported standard file types is limited, and a user may want to
import a non-standard file through the same mechanism. To this end,
the \Rclass{DataTrack} constructor defines an additional argument
called \Rfunarg{importFunction}. As the name suggests, the value of
this argument is a function which needs to handle the mandatory
\Rfunarg{file} argument. Upon evaluation this argument will be filled
in with the path to the data file, and the user-defined function needs
to provide all logic necessary to parse that file into a valid
\Rclass{GRanges} object. From this point on everything will happen just
as if the \Rfunarg{range} argument had been this \Rclass{GRanges}
object. In other words, numeric metadata columns will be
shown as individual samples and non-numeric columns will be silently
ignored. We can exemplify this in the next code chunk. Note that the
\mgg package is using functionality from the \Rpackage{rtracklayer}
package for most of the file import operations, just as we do here in
a more explicit way.

<<filedt2>>=
library(rtracklayer)
dTrack3 <-  DataTrack(range=bgFile, genome="hg19", type="l", chromosome="chr19", name="bedGraph",
                      importFunction=function(file) import(con=file))
identical(dTrack2, dTrack3)
@

So far one could have easily done the whole process in two separate
steps: first import the data from the file into a \Rclass{GRanges}
object and then provided this object to the constructor. The real
power of the file support in the \mgg package comes with streaming
from indexed files. As mentioned before, only the relevant part of the
data has to be loaded during the plotting operation, so the underlying
data files may be quite large without decreasing the performance or
causing too big of a memory footprint. We will exemplify this feature
here using a small \code{bam} file that is provided with the
package. \code{bam} files contain alignments of sequences (typically
from a next generation sequencing experiment) to a common
reference. The most natural representation of such data in a
\Rclass{DataTrack} is to look at the alignment coverage at a given
position only and to encode this in a single metadata column.

<<filedt3, fig=TRUE, width=7.5, height=1>>=
bamFile <- system.file("extdata/test.bam", package="Gviz")
dTrack4 <- DataTrack(range=bamFile, genome="hg19", type="l", name="Coverage", window=-1, chromosome="chr1")
class(dTrack4)
dTrack4
plotTracks(dTrack4, from=189990000, to=190000000)
@

As seen in the previous code chunk, the \code{dTrack4} object is now
of class \Rclass{ReferenceDataTrack}. For the user this distinction is
not particularly relevant with the exception that the
\Rfunction{length} method for this class almost always returns
\code{0} because the content of the object is only realized during the
plotting operation. Obviously, streaming from the disk comes with a
price in that file access is much slower than accessing RAM, however
the file indexing allows for fairly rapid data retrieval, and other
processes during the plotting operation tend to be much more costly,
anyways. It is woth mentioning however that each plotting operation
will cause reading off the file, and there are currently no caching
mechanisms in place to avoid that. Nevertheless, plotting a larger
chunk of the \code{bam} file still finishes in a reasonable time.

<<filedt4, fig=TRUE, width=7.5, height=1>>=
plotTracks(dTrack4, chromosome="chr1", from=189891483, to=190087517)
@

Of course users can provided their own file parsing function just like
we showed in the previous example. The import function now needs to be
able to deal with a second mandatory argument \Rfunarg{selection},
which is a \Rclass{GRanges} object giving the genomic interval that
has to be imported from the file. In addition one needs to tell the
\Rclass{DataTrack} constructor that data should be streamed off a file
by setting the \Rfunarg{stream} argument to \code{TRUE}.

<<filedt5>>=
myImportFun <- function(file, selection){
    ## do something here
}
DataTrack(range=bamFile, genome="hg19", type="l", name="Coverage", window=-1, chromosome="chr1",
          importFunction=myImportFun, stream=TRUE)
@



\subsubsection*{Data transformations}

The \mgg package offers quite some flexibility to transform data on
the fly. This involves both rescaling operations (each data point is
transformed on the track's y-axis by a transformation function) as
well as summarization and smoothing operations (the values for several
genomic locations are summarized into one derived value on the track's
x-axis). To illustrate this let's create a significantly bigger
\Rclass{DataTrack} than the one we used before, containing purely
syntetic data for only a single sample.

<<biggerdata, fig=TRUE, results=hide, width=7.5, height=1.5>>=
dat <- sin(seq(pi, 10*pi, len=500))
dTrack.big <- DataTrack(start=seq(1,100000, len=500), width=15, chromosome="chrX",
                        genome="hg19", name="sinus",
                        data=sin(seq(pi, 5*pi, len=500))*runif(500, 0.5, 1.5))
plotTracks(dTrack.big, type="hist")
@

Since the available resolution on our screen is limited we can no
longer distinguish between individual coordinate ranges. The \mgg
package tries to avoid overplotting by collapsing overlapping ranges
(assuming the \Rfunarg{collapseTracks} parameter is set to
\code{TRUE}). However, it is often desirable to summarize the data,
for instance by binning values into a fixed number of windows followe
by the calculation of a meaningful summary statistic. This can be
archived by a combination of the \Rfunarg{window} and
\Rfunarg{aggregation} display parameters. The former can be an integer
value greater than zero giving the number of evenly-sized bins to
aggregate the data in. The latter is supposed to be a user-supplied
function that accepts a numeric vector as a single input parameter and
returns a single aggregated numerical value. For simplicity, the most
obvious aggregation functions can be selected by passing in a
character scalar rather than a function. Possible values are
\code{mean}, \code{median}, \code{extreme}, \code{sum}, \code{min} and
\code{max}. These presets are also much faster because they have been
optimized to operate on large numeric matrices. The default is to
compute the mean value of all the binned data points.

<<aggregation, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack.big, type="hist", window=50)
@

Instead of binning the data in fixed width bins one can also use the
\Rfunarg{window} parameter to perform more elaborate running window
operations. For this to happen the parameter value has to be smaller
than zero, and the addtional display parameter \Rfunarg{windowSize}
can be used to control the size of the running window. This operation
does not change the number of coordinate ranges on the plot, but
instead the original value at a particular position is replaced by the
respective sliding window value at the same position. A common use
case for sliding windows on genomic ranges is to introduce a certain
degree of smoothing to the data.

<<aggregation2, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack.big, type="hist", window=-1, windowSize=2500)
@

In addition to transforming the data on the x-axis we can also apply
arbitrary transformation functions on the y-axis. One obvious use-case
would be to log-transform the data prior to plotting. The framework is
flexible enough however to allow for arbitrary transformation
operations. The mechanism works by providing a function as the
\Rfunarg{transformation} display parameter, which takes as input a
numeric vector and returns a transformed numeric vector of the same
length. The following code for instance truncates the plotted data to
values greater than zero.

<<transformation, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack.big, type="l", transformation=function(x){x[x<0] <- 0; x})
@

As seen before, the \Rfunarg{a} type allows to plot average values for
each of the separate groups. There is however an additional parameter
\Rfunarg{aggregateGroups} that generalizes group value
aggregations. In the following example we display, for each group and
at each position, the average values in the form of a dot-and-lines
plot.

<<groupingAv1, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack, groups=rep(c("control", "treated"), each=3), type=c("b"), aggregateGroups=TRUE)
@

This functionality again also relies on the setting of the
\Rfunarg{aggregation} parameter, and we can easily change it to
display the maximum group values instead.

<<groupingAv2, fig=TRUE, results=hide, width=7.5, height=1.5>>=
plotTracks(dTrack, groups=rep(c("control", "treated"), each=3), type=c("b"), aggregateGroups=TRUE, aggregation="max")
@

\subsubsection*{Display parameters for DataTrack objects}

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{DataTrack}
class by typing in \code{?DataTrack} on the \RR command line.

<<DataTrackClassTable, echo=FALSE, results=tex>>=
addParTable("DataTrack")
@

\subsection{AnnotationTrack}

\Rclass{AnnotationTrack} objects are the multi-purpose tracks in the
\mgg package. Essentially they consist of one or several genomic
ranges that can be grouped into composite annotation elements if
needed. In principle this would be enough to represent everything from
CpG islands to complex gene models, however for the latter the packge
defines the specialized \Rclass{GeneRegionTrack} class, which will be
highlighted in a separate section. Most of the features discussed here
will also apply to \Rclass{GeneRegionTrack} objects, though. As a
matter of fact, the \Rclass{GeneRegionTrack} class inherits directly
from class \Rclass{AnnotationTrack}.

\Rclass{AnnotationTrack} objects are easily instantiated using the
constructor function of the same name. The necessary building blocks
are the range coordinates, a chromosome and a genome identifier. Again
we try to be flexible in the way this information can be passed to the
function, either in the form of separate function arguments, as
\Rclass{IRanges}, \Rclass{GRanges} or \Rclass{data.frame}
objects. Optionally, we can pass in the strand information for the
annotation features and some useful identifiers. A somewhat special
case is to build the object from a \Rclass{GRangesList} object, which
will automatically preserve the element grouping information contained
in the list structure. For the full details on the constructor
function and the accepted arguments see \code{?AnnotationTrack}. Let's
take a look at a very simple track:

<<anntrack1, fig=TRUE, results=hide, width=7.5, height=0.5>>=
aTrack <- AnnotationTrack(start=c(10, 40, 120), width=15, chromosome="chrX",
                          strand=c("+", "*", "-"),
                          id=c("Huey", "Dewey", "Louie"), genome="hg19", name="foo")
plotTracks(aTrack)
@

The ranges are plotted as simple boxes if no strand information is
available, or as arrows to indicate their direction. We can change the
range item shapes by setting the \Rfunarg{shape} display parameter. It can
also be helpful to add the names for the individual features to the
plot. This can be archived by setting the \Rfunarg{featureAnnotation}
parameter to \code{'id'}.

<<anntrack2, fig=TRUE, results=hide, width=7.5, height=0.5>>=
plotTracks(aTrack, shape="box", featureAnnotation="id")
@

<<anntrack3, fig=TRUE, results=hide, width=7.5, height=0.5>>=
plotTracks(aTrack, shape="ellipse", featureAnnotation="id", fontcolor.feature="darkblue")
@

In this very simplistic example each annotation feature consisted of a
single range. In real life the genomic annotation features that we
encounter often consists of several sub-units. We can create such
composite \Rclass{AnnotationTrack} objects by providing a grouping
factor to the constructor. It needs to be of similar length as the
total number of atomic features in the track, i.e, the number of
genomic ranges that are passed to the constructor. The levels of the
this factor will be used as internal identifiers for the individual
composite feature groups, and we can toggle on their printing by
setting \Rfunarg{groupAnnotation} to \code{'group'}.

<<anntrack4f, fig=TRUE, results=hide, width=7.5, height=0.5>>=
aTrack.groups <- AnnotationTrack(start=c(50, 180, 260, 460, 860, 1240), width=c(15,20,40,100,200, 20),
                                 chromosome="chrX",
                                 strand=rep(c("+", "*", "-"), c(1,3,2)),
                                 group=rep(c("Huey", "Dewey", "Louie"), c(1,3,2)),
                                 genome="hg19", name="foo")
plotTracks(aTrack.groups, groupAnnotation="group")
@

We can control the placement of the group labels through the
\Rfunarg{just.group} parameter.

<<anntrack4af, fig=TRUE, results=hide, width=7.5, height=0.5>>=
plotTracks(aTrack.groups, groupAnnotation="group", just.group="right")
@

<<anntrack4bf, fig=TRUE, results=hide, width=7.5, height=0.5>>=
plotTracks(aTrack.groups, groupAnnotation="group", just.group="above")
@

Arranging items on the plotting canvas is relatively straight forward
as long as there are no overlaps between invidiual regions or groups
of regions. Those inevitably cause overplotting which could seriously
obfuscate the information on the plot. A logical solution to this
problem is to stack overlapping items in separate horizontal lines to
accomodate all of them. This involves some optimization, and the \mgg
package automatically tries to come up with the most compact
arrangement. Let's exemplify this feature with a slightly modified
\Rclass{AnnotationTrack} object.

<<stacking1,  fig=TRUE, results=hide, width=7.5, height=0.5>>=
aTrack.stacked <- AnnotationTrack(start=c(50, 180, 260, 800, 600, 1240), width=c(15,20,40,100,500, 20),
                                 chromosome="chrX",
                                 strand="*",
                                 group=rep(c("Huey", "Dewey", "Louie"), c(1,3,2)),
                                 genome="hg19", name="foo")
plotTracks(aTrack.stacked, groupAnnotation="group")
@

We now have our three annotation feature groups distributed over two
horizontal lines. One can control the stacking of overlapping items
using the \Rfunarg{stacking} display parameter. Currently the three
values \code{squish}, \code{dense} and \code{hide} are
supported. Horizontal stacking is enabled via the \code{squish}
option, which also is the default. The \code{dense} setting forces
overlapping items to be joined in one meta-item and \code{hide} all
together disables the plotting of \Rclass{AnnotationTrack}
items. Please note that adding identifiers to the plot only works for
the \code{squish} option.

<<stacking2,  fig=TRUE, results=hide, width=7.5, height=0.5>>=
plotTracks(aTrack.stacked, stacking="dense")
@

In addition to annotation groups there is also the notion of a feature
type in the \mgg package. Feature types are simply different types of
annotation regions (e.g., mRNA transcripts, miRNAs, rRNAs, etc.) that
are indicated by different colors. There is no limit on the number of
different features, however each element in a grouped annotation item
needs to be of the same feature type. We can query and set features
using the \Rfunction{feature} and \Rfunction{feature<-} methods.

<<features>>=
feature(aTrack.stacked)
feature(aTrack.stacked)<- c("foo", "bar", "bar", "bar", "no", "no")
@

Features types can also be shown as either the group or the item
annotation on the plot.

<<featuresIdPlot, fig=TRUE, results=hide, width=7.5, height=0.5>>=
plotTracks(aTrack.stacked, featureAnnotation="feature", groupAnnotation="feature", fontcolor.feature=1, cex.feature=0.7)
@

Unless we tell the \mgg package how to deal with the respective
feature types they will all be treated in a similar fashion, i.e.,
they will be plotted using the default color as defined by the
\Rfunarg{fill} display paramter. To define colors for individual
feature types we simply have to add them as additional display
parameters, where the parameter name matches the feature type and its
value is supposed to be a valid R color qualifier. Of course this
implies that we can only use feature names that are not already
occupied by other display parameters defined in the package.


<<featuresPlotf, fig=TRUE, results=hide, width=7.5, height=0.5>>=
plotTracks(aTrack.stacked, groupAnnotation="group", foo="darkred", bar="darkgreen")
@

Stacking of annotation items to avoid overplotting only works as long
as there is enough real estate on the plotting canvas to separate all
items, i.e., we need all items to be at least a single pixel wide to
correctly display them. This limitation is automatically enforced by
the \mgg package, however it implies that unless neighbouring items
are more than one pixel appart we can not distinguish between them and
will inevitably introduce a certain amount of overplotting. This means
that on a common screen device we can only look at a very limited
genomic region of a few kb in full resolution. Given that an average
chromosome is in the order of a few gb, we still need a reasonable way
to deal with the overplotting problem despite the item stacking
functionality. As default, the \mgg package will merge all overlapping
items into one unified meta-item and only plot that (see 'Collapse'
section below for details). In order to indicate the amount of
overplotting that was introduced by this process we can use the
\Rfunarg{showOverplotting} display parameter. It uses a color scale
(based on the orginal colors defined for the track), with lighter
colors indicating areas of low or no overplotting, and more saturated
colors indicating areas of high overplotting density. We exemplify
this feature on an \Rclass{AnnotationTrack} object that represents a
good portion of a real human chromosome.

<<overplotting,  fig=TRUE, results=hide, width=7.5, height=0.75>>=
data("denseAnnTrack")
plotTracks(denseAnnTrack, showOverplotting=TRUE)
@



\subsubsection*{Collapsing}
All track types that inherit from class \code{AnnotationTrack} support
the collapsing of overlapping track items, either because they have
initially been defined as overlapping coordinates, or because the
current device resolution does not allow to sufficiently separate
them. For instance, two elements of a feature group may be separated
by 100 base pairs on the genomic scale, however when plotted to the
screen, those 100 base pairs translate to a distance of less than one
pixel. In this case we can no longer show the items as two separate
entitites. One solution to this problem would be to allow for
arbitrary overplotting, in which case the last one of the overlapping
items that is drawn on the device wins. This is not optimal in many
ways, and it also poses a significant burden on the graphical engine
because a lot of stuff has to be drawn which no one will ever see.

To this end the \mgg package provides an infrastructure to reasonably
collapse overlappig items, thereby adjusting the information content
that can be shown to the available device resolution. By default this
feature is turned on, and the user does not have to worry too much
about it. However, one should be aware of the consequences this may
have on a given visualization. If you absolutely do not want
collapsing to take place, you may completely turn it off by setting
the display parameter \code{collapse} to \code{FALSE}. Please note
that by doing this the \code{showOverplotting} parameter will also
stop working. If you opt in, there is some considerable amount of
detailed control to fine tune the collapsing to your needs.

Let's start with a small example track for which element collapsing has
been turned off and no adjustments to the ranges have been made. We
plot both the item identifiers and the group identifiers to exemplify
what is going on.

<<collapse1f, fig=TRUE, results=hide, width=7.5, height=0.85>>=
data(collapseTrack)
plotTracks(ctrack)
@

The first thing to notice is that the for item \code{d} we do see the
item identifier but not the range itself. This is due to the fact that
the width of the item is smaller than a single pixel, and hence the
graphics system can not display it (Note that this is only true for
certain devices. The quartz device on the Mac seems to be a little
smarter about this). There are also the two items \code{e} and
\code{f} which seem to overlay each other completely, and another two
items which appear to be just a single fused  item (\code{k} and
\code{l}). Again, this is a resolution issue as their relative
distance is smaller than a single pixel, so all we see is a single
range and some ugly overplotted identifiers. We can control the first
issue by setting the minimum pixel width of a plotted item to be one
pixel using the \code{min.width} display parameter.

<<collapse2f, fig=TRUE, results=hide, width=7.5, height=0.85>>=
plotTracks(ctrack, min.width=1)
@

Now the item \code{d} has a plotable size and can be drawn to the
device. The overplotted items are still rather anoying, but the only way to
get rid of those is to turn item collapsing back on.

<<collapse3f, fig=TRUE, results=hide, width=7.5, height=0.85>>=
plotTracks(ctrack, min.width=1, collapse=TRUE)
@

Now all items that could not be separated by at least one pixel have
been merged into a single meta-item, and the software has taken care
of the identifiers for you, too. The merging operation is aware of the
grouping information, so no two groups where joint together. Sometimes
a single pixel width or a single pixel distance is not enough to get a
good visualization. In these cases one could decide to enforce even
larger values. We can do this not only for the minimum width, but
also for the minimum distance by setting the \code{min.distance}
parameter.

<<collapse4f, fig=TRUE, results=hide, width=7.5, height=0.85>>=
plotTracks(ctrack, min.width=3, min.distance=5, collapse=TRUE)
@

This time also the two items \code{b} and \code{c} have been merged,
and all ranges are now at least 3 pixels wide. Depending on the
density of items on the plot even this reduction can be
insufficient. Because we did not merge complete groups we might still
end up with quite a lot of stacks to accomodate all the
information. To this end the display parameter \code{mergeGroups} can
be used to disable absolute group separation. Rather than blindly
merging all groups (as it is done when \code{stacking='dense'})
however, the software will only join those overlapping group ranges
for which all items are already merged into a single meta item.

<<collapse5f, fig=TRUE, results=hide, width=7.5, height=0.65>>=
plotTracks(ctrack, min.width=3, min.distance=5, collapse=TRUE,
           mergeGroups=TRUE, extend.left=0.1)
@

\subsubsection*{Building AnnotationTrack objects from files}
Just like we showed before for \Rclass{DataTrack} objects it is
possible to stream the data for an \Rclass{AnnotationTrack} object off
the file system. We will again use the \code{bam} file as an
example. This time the default import function reads the coordinates
of all the sequence alignments from the file and even knows how to
deal with grouped elements.

<<fileat1, fig=TRUE, width=7.5, height=1>>=
aTrack2 <- AnnotationTrack(range=bamFile, genome="hg19", name="Reads", chromosome="chr1")
class(aTrack2)
aTrack2
plotTracks(aTrack2, from=189995000, to=190000000)
@

Since \Rclass{AnnotationTrack} objects are somewhat more complicated
than \Rclass{DataTrack} objects, the constructor provides another
level of flexibility when reading the data from a file. More
specifically, the user has the ability to map the provided
metadata columns of the \Rclass{GRanges} object that is
returned by the import function to the respective columns in the final
\Rclass{AnnotationTrack} object. If no explicit mapping is provided,
the package will try to find a default mapping based on the input file
type and the desired track type. The current mapping of a track is
displayed by its \Rfunction{show} method as can be seen above. The
mechanism for changing this default mapping is quite straight forward:
if the values in the relevant constructor arguments (in this case
\Rfunarg{group}, \Rfunarg{id} and \Rfunarg{feature}) are character
scalars, the metadata columns of the \Rclass{GRanges}
object that is returned by the import function are mapped against
their values. For instance we can map the \code{id} column to the
\code{group} attribute and thus suppress the grouping of sequence
alignments that originate from the same sequencing read.

<<fileat2, fig=TRUE, width=7.5, height=1>>=
aTrack3 <- AnnotationTrack(range=bamFile, genome="hg19", name="Reads", chromosome="chr1", group="id")
aTrack3
plotTracks(aTrack3, from=189995000, to=190000000)
@

Please note that if no metadata column can be mapped to the
constructor arguments their values are assumed to be normal character
scalars and they will be recycled accordingly to fill the track's
slots. One can ask for the package's default mappings by using the
\Rfunction{availableDefaultMapping} function.

<<fileat3>>=
availableDefaultMapping(bamFile, "AnnotationTrack")
@

We can now plot both the \Rclass{DataTrack} representation as well as
the \code{AnnotationTrack} representation of the \code{bam} file
together to prove that the underlying data are indeed identical.

<<fileat4, fig=TRUE, width=7.5, height=2>>=
plotTracks(list(dTrack4, aTrack2), from=189990000, to=190000000)
@


\subsubsection*{Display parameters for AnnotationTrack objects}

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{AnnotationTrack}
class by typing in \code{?AnnotationTrack} on the \RR command line.

<<AnnotationTrackClassTable, echo=FALSE, results=tex>>=
addParTable("AnnotationTrack")
@

\subsection{GeneRegionTrack}

\Rclass{GeneRegionTrack} objects are in principle very similar to
\Rclass{AnnotationTrack} objects. The only difference is that they are
a little more gene/transcript centric, both in terms of plotting
layout and user interaction, and that they may define a global start
and end position (a feature which is not particularly relevant for the
normal user). The constructor function of the same name is a
convenient tool to instantiate the object from a variety of different
sources. In a nutshell, we need to pass start and end positions (or
the width) of each annotation feature in the track and also supply the
exon, transcript and gene identifiers for each item which will be used
to create the transcript groupings. A somewhat special case is to
build a \Rclass{GeneRegionTrack} object directly from one of the
popular \Rclass{TxDb} objects, an option that is treated in
more detail below. For more information about all the available
options see the class's manual page (\code{?GeneRegionTrack}).

There are a number of accessor methods that make it easy to query and
replace for instance exon, transcript or gene assignments. There is
also some support for gene aliases or gene symbols which are often
times more useful than cryptic data base gene identifiers. The
following code that re-uses the \Rclass{GeneRegionTrack} object from
the first section exemplifies some of these features.

<<generegtrackf,  fig=TRUE, results=hide, width=7.5, height=1.5>>=
data(geneModels)
grtrack <- GeneRegionTrack(geneModels, genome=gen, chromosome=chr, name="foo")
head(gene(grtrack))
head(transcript(grtrack))
head(exon(grtrack))
head(symbol(grtrack))
plotTracks(grtrack)
@

Both exon- and transcript-centric annotation can be added to the plot
using the \Rfunarg{transcriptAnnotation} and \Rfunarg{exonAnnotation}
display parameters.

<<generegtrack2af,  fig=TRUE, results=hide, width=7.5, height=2.5>>=
plotTracks(grtrack, transcriptAnnotation="symbol")
@

<<generegtrack2bf,  fig=TRUE, results=hide, width=7.5, height=2.5>>=
plotTracks(grtrack, transcriptAnnotation="transcript")
@

<<generegtrack2bf,  fig=TRUE, results=hide, width=7.5, height=1>>=
plotTracks(grtrack, exonAnnotation="exon", extend.left=-0.8, fontcolor.exon=1)
@

Since we have the gene and trancript level information as part of our
\Rclass{GeneRegionTrack} objects we can ask the package to collapse
all of our gene models from individual exons and transcripts down to
gene body locations by setting the \Rfunarg{collapseTranscripts}
display parameter to \code{TRUE}.

<<generegtrack3f, fig=TRUE, results=hide, width=7.5, height=1>>=
plotTracks(grtrack, collapseTranscripts=TRUE, shape="arrow", transcriptAnnotation="symbol")
@

Collapsing down all the way to a gene is not always the desired
oparation, and the parameters offers a bit more control. For instance
we could decide to plot just the longest transcript by setting it to
\code{longest}.

<<generegtrack3g, fig=TRUE, results=hide, width=7.5, height=1>>=
plotTracks(grtrack, collapseTranscripts="longest", shape="arrow", transcriptAnnotation="symbol")
@

Or we can plot the union of all the exons in some sort of meta-transcript.

<<generegtrack3h, fig=TRUE, results=hide, width=7.5, height=1>>=
plotTracks(grtrack, collapseTranscripts="meta", shape="arrow", transcriptAnnotation="symbol")
@

\subsubsection*{Building GeneRegionTrack objects from TxDbs}

The \Rpackage{GenomicFeatures} packages provides an elegant framework
to download gene model information from online sources and to store it
locally in a SQLite data base. Because these so called
\Rclass{TxDb} objects have become the de-facto standard for
genome annotation information in Bioconductor we tried to make it as
simple as possible to convert them into
\Rclass{GeneRegionTracks}. Essentially one only has to call the
constructor function with the \Rclass{TxDb} object as a single
argument. We exemplify this on a small sample data set that comes with
the \Rpackage{GenomicFeatures} package.

<<tdb2grt1>>=
library(GenomicFeatures)
samplefile <- system.file("extdata", "hg19_knownGene_sample.sqlite", package="GenomicFeatures")
txdb <- loadDb(samplefile)
GeneRegionTrack(txdb)
@

In this context, the constructor's \Rfunarg{chromosome},
\Rfunarg{start} and \Rfunarg{end} argument take on a slightly differnt
meaning in that they can be used to subset the data that is fetched
from the \Rclass{TxDb} object. Please note that while the
\Rfunarg{chromosome} alone can be supplied, providing \Rfunarg{start}
or \Rfunarg{end} without the chromosome information will not work.

<<tdb2grt2>>=
txTr <- GeneRegionTrack(txdb, chromosome="chr6", start=35000000, end=40000000)
@

A nice bonus when building \Rclass{GeneRegionTracks} from
\Rclass{TxDb} objects is that we get additional information
about coding and non-coding regions of the transcripts, i.e.,
coordinates of the 5' and 3' UTRs and of the CDS regions. The class'
plotting method can use this inforamtion to distinguish between coding
and non-coding regions based on the shape of the feature. All coding
regions are plotted just as we have seen in the previous examples,
whereas the non-coding regions are drawn as slighly thinner boxes. The
distinction between coding and non-coding is made on the basis of the
object's \code{feature} values in combination with a special display
parameter \code{thinBoxFeature} that enumerates all feature types that
are to be treated as non-coding. Obviously this feature is available
to all \Rclass{GeneRegionTracks}, not only the ones that were build
from \Rclass{TxDb} objects. However, the coding information
has to be added manually and the default value of the
\code{thinBoxFeature} parameter may not be sufficient to cover all
possible cases. It is up to the user to come up with a complete
list of non-coding feature types depending on the source of the data.

<<generegtrack4f, fig=TRUE, results=hide, width=7.5, height=0.35>>=
feature(txTr)
plotTracks(txTr)
@

\subsubsection*{Building GeneRegionTrack objects from EnsDbs}

An alternative resource of gene annotations is provided by annotation packages
built by the \Rpackage{ensembldb} package. The \Rclass{EnsDb} object from these
packages are similar in functionality than \Rclass{TxDb} objects, but provide,
and are tailored for, gene annotations from Ensembl. In addition, the Ensembl
based annotation provides gene and transcript biotypes and the \Rclass{EnsDb}
allows to further filter results based on its filter system. In the code below
we fetch all genes defined in Ensembl version 75 within the region defined by
the example \code{geneModels}.

Importantly, we have to set the \code{"ucscChromosomeNames"} option to
\code{FALSE} in order to work with Ensembl based annotations.

<<ensdb2grt1>>=
library(EnsDb.Hsapiens.v75)
ensdb <- EnsDb.Hsapiens.v75

## Disabling the automatic renaming of chromosome names to UCSC format!
options(ucscChromosomeNames=FALSE)

grtr <- GeneRegionTrack(ensdb, start=min(geneModels$start), end=max(geneModels$end),
                        chromosome=sub(unique(geneModels$chromosome), pattern="chr",
                                       replacement=""))
@

<<ensdb2grtplot, fig=TRUE, results=hide, width=7.5, height=1.25>>=
plotTracks(grtr, transcriptAnnotation="symbol")
@

Alternatively, we could fetch e.g. lincRNAs and protein coding genes separately.

<<ensdb2grt2>>=
lincRNAs <- GeneRegionTrack(ensdb, start=min(geneModels$start), end=max(geneModels$end),
                            chromosome=sub(unique(geneModels$chromosome), pattern="chr",
                                           replacement=""),
                            filter=GenebiotypeFilter("lincRNA"), name="lincRNAs")
protCoding <- GeneRegionTrack(ensdb, start=min(geneModels$start), end=max(geneModels$end),
                              chromosome=sub(unique(geneModels$chromosome), pattern="chr",
                                             replacement=""),
                              filter=GenebiotypeFilter("protein_coding"),
                              name="proteins")

@

<<ensdb2grtplot2, fig=TRUE, results=hide, width=7.5, height=2.25>>=
plotTracks(list(lincRNAs, protCoding), transcriptAnnotation="symbol")
@

At last we re-set the UCSC chromosme name option to \code{TRUE} so that we don't
get any errors in the following chunks.

<<ensdbreset>>=
options(ucscChromosomeNames=TRUE)
@

\subsubsection*{Display parameters for GeneRegionTrack objects}

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{GeneRegionTrack}
class by typing in \code{?GeneRegionTrack} on the \RR command line.

<<GeneRegionTrackClassTable, echo=FALSE, results=tex>>=
addParTable("GeneRegionTrack")
@

\subsection{BiomartGeneRegionTrack}

As seen before it can be very useful to quickly download gene
annotation information from an online repositry rather than having to
construct it each time from scratch. To this end, the \mgg package
also defines the \Rclass{BiomartGeneRegionTrack} class, which directly
extends \Rclass{GeneRegionTrack} but provides a direct interface to
the ENSEMBL Biomart service (yet another interface to the UCSC data
base content is highlighted in one of the next sections). Rather than
providing all the bits and pieces for the full gene model, we just
enter a genome, chromosome and a start and end position on this
chromosome, and the constructor function
\Rfunction{BiomartGeneRegionTrack} will automatically contact ENSEMBL,
fetch the necessary information and build the gene model on the
fly. Please note that you will need an internet connection for this to
work, and that contacting Biomart can take a significant amount of
time depending on usage and network traffic. Hence the results are
almost never going to be returned instantaniously.

<<BiomartGeneRegionTrackShow, eval=FALSE>>=

biomTrack <- BiomartGeneRegionTrack(genome="hg19", chromosome=chr, start=20e6, end=21e6,
                                  name="ENSEMBL")
plotTracks(biomTrack)
@
<<BiomartGeneRegionTrackDo, echo=FALSE, results=hide, width=7.5, height=1.25, fig=TRUE>>=
if(hasBiomartConnection){
    biomTrack <- BiomartGeneRegionTrack(genome="hg19", chromosome=chr, start=20e6, end=21e6,
                                        name="ENSEMBL")
}else{
    data("biomTrack")
}
plotTracks(biomTrack)
@

You may have noticed in the above plot that the track includes
\code{feature} information which is displayed by the different feature
colors and box sizes. This information has been automatically
extracted from Biomart and the respective color coding is part of the
class' definition. We can highlight the feature classes even more by
using similarly colored bars to connect the grouped elements and by
dropping the bounding boxes around the individual exon models.

<<BiomartGeneRegionTrackCol, width=7.5, height=1.25, fig=TRUE>>=
plotTracks(biomTrack, col.line=NULL, col=NULL)
@

For dense stacked plots like this it can sometimes be helpful to
control the vertical spacing between the stacking rows. A simple way
to do this is available for each object inheriting form the
\Rclass{StackedTrack} class by means of the \Rfunarg{stackHeight}
display parameter. It takes a value between 0 and 1 that controls
which fraction of the available vertical space on a stacking line
should be used to draw the feature glyphs.

<<BiomartGeneRegionTrackHeight, width=7.5, height=1.25, fig=TRUE>>=
plotTracks(biomTrack, col.line=NULL, col=NULL, stackHeight=0.3)
@

The \Rfunarg{filter} argument to the
\Rfunction{BiomartGeneRegionTrack} constructor can be used to pass on
arbitratry additional filters to the Biomart query. We could for
instance limit the returned gene models to RefSeq models only. Please
see the documentation in the \Rpackage{biomaRt} package for details.


<<BiomartGeneRegionTrackFilterShow, eval=FALSE>>=
biomTrack <- BiomartGeneRegionTrack(genome="hg19", chromosome=chr, start=20e6, end=21e6,
                                  name="ENSEMBL", filter=list(with_ox_refseq_mrna=TRUE))
plotTracks(biomTrack, col.line=NULL, col=NULL, stackHeight=0.3)
@
<<BiomartGeneRegionTrackFilterDo, width=7.5, height=1.25, fig=TRUE, echo=FALSE, results=hide>>=
if(hasBiomartConnection){
    biomTrack <- BiomartGeneRegionTrack(genome="hg19", chromosome=chr, start=20e6, end=21e6,
                                  name="ENSEMBL", filter=list(with_ox_refseq_mrna=TRUE))
    plotTracks(biomTrack, col.line=NULL, col=NULL, stackHeight=0.3)
}else{
    biomTrack@filter <- list(with_ox_refseq_mrna=TRUE)
    plotTracks(biomTrack, col.line=NULL, col=NULL, stackHeight=0.3)
    biomTrack@filter <- list()
}
@

Sometimes it can be more convenient to get a model for a particular
gene rather than defining a genomic range. One could use the filter
mechanisms as described before for this purpose, however the
\Rclass{BiomartGeneRegionTrack} implements a more generic way. By
providing the gene symbol as the \Rfunarg{symbol} argument, the track
will be initialized around the locus for this gene, if it can be
found. Alternatively, one can provide the Ensembl trancript or gene id
via the  \Rfunarg{transcript} or \Rfunarg{gene} arguments, or an Entrez id via
the \Rfunarg{entrez} argument. In all of these cases the constructor
will also fetch models for other genes that overlap the same
region. If that is not desired one needs to provide an explicit
Biomart filter.

<<BiomartGeneRegionTrackSymbolShow, eval=FALSE>>=
biomTrack <- BiomartGeneRegionTrack(genome="hg19", name="ENSEMBL", symbol="ABCB5")
plotTracks(biomTrack)
@
<<BiomartGeneRegionTrackSymbolDo, width=7.5, height=1.25, fig=TRUE, echo=FALSE, results=hide>>=
if(hasBiomartConnection){
    biomTrack <- BiomartGeneRegionTrack(genome="hg19", name="ENSEMBL", symbol="ABCB5")
    plotTracks(biomTrack, transcriptAnnotation="symbol")
}else{
    ranges(biomTrack) <- ranges(biomTrack)[symbol(biomTrack) == "ABCB5"]
    plotTracks(biomTrack, transcriptAnnotation="symbol")
}
@

\subsubsection*{Using custom biomaRt objects for BiomartGeneRegionTracks}
\mgg tries hard to automatically set up the necessary connection to
the Biomart service to retrieve the relevant data. Sometimes however
it can be advantagous to fully control the Biomart connector, for
instance when trying to retrieve data from an archived Ensembl version
(by default, \Rpackage{biomaRt} will always link to the most recent
version). To that end, the \Rfunarg{biomart} parameter can be used to
supply a \Robject{biomaRt} object. Please note that a data set has to
be selected in that object. A second prerequesite for this to work is
that \mgg is able to guess which of the values that are queried
from the Ensembl data base represent which track feature. If there
were no changes in the schema of Ensembl, or if the schema for all
marts was similar this would not be an issue, but that is not the
case. In the following example we show how to fetch data for the same
region from the older version Ensembl version 47. The internal
\Rfunction{.getBMFeatureMap} is helpful to set up the feature mapping
which is necessary because Esembl used to call the gene symbol field
\code{external\_gene\_id} in that release.

<<BiomartGeneRegionTrackCustom, width=7.5, height=1.25, fig=TRUE>>=
library(biomaRt)
bm <- useMart(host="oct2012.archive.ensembl.org", biomart="ENSEMBL_MART_ENSEMBL",
              dataset="hsapiens_gene_ensembl")
fm <- Gviz:::.getBMFeatureMap()
fm["symbol"] <- "external_gene_id"
biomTrack <- BiomartGeneRegionTrack(genome="hg19", chromosome="chr7", start=20e6, end=21e6,name="ENSEMBL",
                                    featureMap=fm, biomart=bm)
plotTracks(biomTrack, col.line=NULL, col=NULL, stackHeight=0.3)
@


\subsubsection*{Display parameters for BiomartGeneRegionTrack objects}

For a complete listing of all the available display parameters please
see the table above in the previous \Rclass{GeneRegionTrack} section
or the man page of the \Rclass{BiomartGeneRegionTrack} class by typing
in \code{?BiomartGeneRegionTrack} on the \RR command line.

As mentioned above, one additional benefit when fetching the data
through Biomart is that we also receive some information about the
annotation feature types, which is automatically used for the color
coding of the track. The following table shows the available feature
types.

<<BiomartGeneRegionTrackClassTable, echo=FALSE, results=tex>>=
addInfo <- t(data.frame(displayPars(biomTrack, names(details[["BiomartGeneRegionTrack"]]))))
colnames(addInfo) <- "Color"
addParTable("BiomartGeneRegionTrack", add=addInfo)
@

\subsection{DetailsAnnotationTrack}
It is sometimes desirable to add more detailed information to
particular ranges in an \mgg plot for which the notion of genomic
coordinates no longer makes sense. For instance, the ranges in an
\Rclass{AnnotationTrack} may represent probe locations on a genome,
and for each of these probes a number of measurements from multiple
samples and from different sample groups are available. To this end,
the \Rclass{DetailsAnnotationTrack} provides a flexible interface to
further annotate genomic regions with arbitrary additional
information. This is archived by splitting the
\Rclass{AnnotationTrack} plotting region into two horizontal sections:
the lower section containing the range data in genomic coordinates, and
the upper one containing the additional data for each of the displayed
ranges in verticaly tiled panels of equal size. The connection between
a range item and its details panel is indicated by connecting
lines.

The content of the individual details panels has to be filled in by a
user-defined plotting function that uses grid (or lattice) plotting
commands. This function has to accept a number of mandatory
parameters, notably the start, end, strand, chromosome and identifier
information for the genomic range, as well as an integer counter
indicating the index of the currently plotted details tile. This
information can be used to fetch abtritray details, e.g. from a list,
and environement or even from a \Rclass{GRanges} object which will
then be processed and visualized within the plotting function. This
may sound rather abstract, and for more details please refer to the
class' help page. For now we just want to demonstrate the
functionality in a simple little example. We begin by defining a
\Rclass{GRanges} object containing 4 genomic locations. In our example
those are considered to be probe locations from a methylation array.

<<DetailsAnnotationTrack1>>=
library(GenomicRanges)
probes <- GRanges(seqnames="chr7", ranges=IRanges(start=c(2000000, 2070000, 2100000, 2160000), end=c(2050000, 2130000, 2150000, 2170000)),
                  strand=c("-", "+", "-", "-"))
@

For each of these probes we have methylation measurements from a large
number of different samples in a numeric matrix, and within the
samples there are two treatment groups. The aim is to compare the
distribution of measurement values between these two groups at each
probe locus.

<<DetailsAnnotationTrack2>>=
methylation <- matrix(c(rgamma(400, 1)), ncol=100, dimnames=list(paste("probe", 1:4, sep=""), NULL))
methylation[,51:100] <- methylation[,51:100] + 0:3
sgroups <- rep(c("grp1","grp2"), each=50)
@

Of course we could use a \Rclass{DataTrack} with the box-plot
representation for this task, however we do have strand-specific data
here and some of the probes can be overlapping, so all this
information would be lost. We are also interested in the particular
shape of the data distribution, so a density plot representation is
what we really need. Luckily, the \Rpackage{lattice} package gives us
a nice \Rfunction{densityplot} function that supports grouping, so all
that's left to do now is to write a little wrapper that handles the
extraction of the relevant data from the matrix. This is easily
archieved by using the range identifiers, which conveniently map to
the row names of the data matrix.

<<DetailsAnnotationTrack3>>=
library(lattice)
details <- function(identifier, ...) {
    d <- data.frame(signal=methylation[identifier,], group=sgroups)
    print(densityplot(~signal, group=group, data=d, main=list(label=identifier, cex=0.7),
                      scales=list(draw=FALSE, x=list(draw=TRUE)), ylab="", xlab="",
                      ), newpage=FALSE, prefix="plot")
}
@

Finaly, it is as simple as calling the \Rclass{AnnotationTrack}
constructor, passing along the wrapper function and calling
\Rfunction{plotTracks}.

<<DetailsAnnotationTrack4, results=hide, width=7.5, height=5, fig=TRUE>>=
deTrack <- AnnotationTrack(range=probes, genome="hg19", chromosome=7, id=rownames(methylation),
                           name="probe details", stacking="squish",
                           fun=details)
plotTracks(deTrack)
@

It should be noted here that in our little example we rely on the
methylation data matrix and the grouping vector to be present in the
working environment. This is not necessarily the cleanest solution and
one should consider storing additional data in an evironment, passing
it along using the \Rfunarg{detailFunArgs} parameter, or making it
part of the details function in form of a closure. The class' help
page provides further instructions.

Another use case for the \Rclass{DetailsAnnotationTrack} class is to
deal with the problem of very different feature sizes within a single
track. For instance, we may be looking at a rather large genomic
region containing one big transcript with many widely spaced exons and
a bunch of smaller, more compact transcripts. In this case it would be
helpful to provide a zoomed in version of those smaller
transcripts. In order to achieve this we can make use of the class'
\code{groupDetails} display parameter, which applies the detail
plotting function over each range group rather than over individual
range items. First we define a function that selects those groups with
a plotted size smaller than 10 pixels. We make use of the unexported
function \code{.pxResolution} here to come up with the mapping between
pixel coordinates and genomic coordinates.

<<DetailsAnnotationTrack5>>=
selFun <- function(identifier, start, end, track, GdObject, ...){
    gcount <- table(group(GdObject))
    ## This computes the width of 2 pixels in genomic coordinates
    pxRange <- Gviz:::.pxResolution(min.width=20, coord="x")
    return((end-start)<pxRange && gcount[identifier]==1)
}
@

The actual detail plotting function is fairly trivial. In the details
viewport we simply call \code{plotTracks} function on the subset of
items from each group that has been selected before without plotting
the track titles, and also add a scale indicator for some
reference. It is worth mentioning however that we need to extract the
ranges of the zoomed in track items from the unmodified track object
(the \code{Gdobject.original} argument in the details function)
because the object that is plotted in the lower panel contains only
the collapsed ranges.

<<DetailsAnnotationTrack6>>=
detFun <- function(identifier, GdObject.original, ...){
    plotTracks(list(GenomeAxisTrack(scale=0.3, size=0.2, cex=0.7), GdObject.original[group(GdObject.original)==identifier]),
               add=TRUE, showTitle=FALSE)
}
@

Finally, we load some sample data, turn it into a \code{DetailsAnnotationTrack} object and plot it.

<<DetailsAnnotationTrack7f, results=hide, width=7.5, height=2, fig=TRUE>>=
data(geneDetails)
deTrack2 <- AnnotationTrack(geneDetails, fun=detFun, selectFun=selFun,
                            groupDetails=TRUE, details.size=0.5, detailsConnector.cex=0.5, detailsConnector.lty="dotted",
                            shape=c("smallArrow", "arrow"), groupAnnotation="group")
plotTracks(deTrack2, extend.left=90000)
@


\subsubsection*{Display parameters for DetailsAnnotationTrack objects}

In addtion to the display parameters for the \Rclass{AnnotationTrack}
class, some additional parameters can be used to control the look and
feel of the details sections. For a complete listing of all the available display parameters please
see the tables below and the one above in the \Rclass{AnnotationTrack} section
or the man page of the \Rclass{DetailsAnnotationTrack} class by typing
in \code{?DetailsAnnotationTrack} on the \RR command line.

<<DetailsAnnotationTrack5, results=hide, width=7.5, height=3, fig=TRUE>>=
plotTracks(deTrack, details.size=0.75, detailsConnector.pch=NA, detailsConnector.col="darkred",
           detailsBorder.fill="#FFE3BF", detailsBorder.col="darkred", shape="box", detailsConnector.lty="dotted")
@

<<DetailsAnnotationTrackClassTableSec, echo=FALSE, results=tex>>=
addParTable("DetailsAnnotationTrack")
@


\subsection{SequenceTrack}
So far we have displayed all kinds of ranges on a genome, but never really
looked at the underlying genomic sequence. To that end, the \mgg package defines
the \Rclass{SequenceTrack} class which can hold genomic sequence information
derived from either a \Rclass{DNAStringSet} object of, even more convenient,
from one of the \Rclass{BSgenome} packages. Alternatively, a \Rclass{FaFile} as
provided by the \Rpackage{AnnotationHub} package can be used as input to the
\Rclass{SequenceTrack} constructor. There is not formal checking of the track's
genome, so it is up to the user to provide reasonable inputs to the class'
constructor. As with all the other track types, the constructor has the same
name as the class. In this example we build a \Rclass{SequenceTrack} from the
human hg19 UCSC genome.

<<SequenceTrack1>>=
library(BSgenome.Hsapiens.UCSC.hg19)
sTrack <- SequenceTrack(Hsapiens)
sTrack
@

Here we retain the same benefits as for regular \Rclass{BSgenome}
objects in that the actual sequence is only loaded into memory when a
particular chromosome is first accessed. Plotting the track yields the
expected result: a character representation of the genomic sequence in
the current plot window. It may be worth noting that the color scheme
that is used to encode the indvidual nucleotides is defined in the
\Rpackage{biovizBase} package and the user is referred to its
documentation for more details.

<<SequenceTrack2, results=hide, width=7.5, height=0.5, fig=TRUE>>=
plotTracks(sTrack, chromosome=1, from=20000, to=20050)
@

Sometimes it may be too busy to plot both the sequence letters and the
color coding, and we can assign arbitrary colors by setting the
\code{fontcolor} display parameter. The convention here is that we
need to supply a named vector, with one entry for each of the five
possible nucleotides (A, C, T, G, and N). If any of the entries is
missing, the respective letter will not be drawn at all.

<<SequenceTrack3, results=hide, width=7.5, height=0.5, fig=TRUE>>=
fcol <- c(A="darkgray", C="darkgray", T="darkgray", G="darkgray")
plotTracks(sTrack, chromosome=1, from=20000, to=20050, fontcolor=fcol)
@

In this case we are plotting the sequence on the forward strand and we
may want to make this clear by adding direction indicators.

<<SequenceTrack4, results=hide, width=7.5, height=0.5, fig=TRUE>>=
plotTracks(sTrack, chromosome=1, from=20000, to=20050, add53=TRUE)
@

If instead we plot the complement sequence on the reverse strand, the
indicators are automatically adjusted.

<<SequenceTrack5, results=hide, width=7.5, height=0.3, fig=TRUE>>=
plotTracks(sTrack, chromosome=1, from=20000, to=20050, add53=TRUE, complement=TRUE)
@

So far we have been able to fit the sequence onto our plotting device
without overplotting. I.e., all the letters nicely fit in to their
respective position in the genomic coordinate system. However this
restricts us to very small windows which we can reasonably display. A
much more compact version of the same information is to use colored
boxes rather than individual letters. Those boxes can be stacked much
closer together, which increases the possible window size quite a
bit. The user does not really need to worry about this as the plotting
method will automatically make a reasonable decision based on the
available space.

<<SequenceTrack6, results=hide, width=7.5, height=0.3, fig=TRUE>>=
plotTracks(sTrack, chromosome=1, from=20000, to=20100)
@

For added flexibility one can set the \code{noLetters} display
paramter to \code{TRUE} to always force the drawing of boxes.  Of
course also the colored box represenation of a sequence has its
limits, and if we cross the threshold when individual boxes can not be
separated anymore, the plotting method falls back to drawing a single
line indicating the presence of a sequence at the given position.

<<SequenceTrack7, results=hide, width=7.5, height=0.3, fig=TRUE>>=
plotTracks(sTrack, chromosome=1, from=20000, to=201000)
@

Finally, the selected font size is also contributing to the available
space, and we can cram a little more sequence into a given window by
decreasing it.

<<SequenceTrack8, results=hide, width=7.5, height=0.3, fig=TRUE>>=
plotTracks(sTrack, chromosome=1, from=20000, to=20100, cex=0.5)
@

An alternative method to create a \code{SequenceTrack} is by submitting directly
a \code{FaFile} to the constructor. The \code{AnnotationHub} package provides
such files, e.g. representing genomic sequence fasta files from Ensembl.
In the code below we query and fetch a \code{FaFile} containing the genomic
sequence for Homo sapiens from Ensembl release 80 (thus representing GRCh38)
from \code{AnnotationHub}. Note that the code is not executed during generation
of this vignette due to required download of the large fasta file. The
\code{AnnotationHub} however caches downloads locally, thus subsequent calls
to the same resource would not require any additional downloads.

<<AnnotationHub1, results=hide, eval=FALSE>>=
## Initializing AnnotationHub ressource.
library(AnnotationHub)
ah <- AnnotationHub()
## Query all available files for Ensembl release 80 and Homo sapiens.
query(ah, c("Homo sapiens", "release-80"))

## One of the entries represents the "dna.toplevel" fasta file
## containing the full genomic sequence. We will fetch that file.
dna <- ah[["AH47266"]]
@

Now that we have the genomic sequence we can create the corresponding
\code{SequenceTrack}. Since we're working here with Ensembl annotations, we have
to set the \code{"ucscChromosomeNames"} option to false (i.e. Ensembl names
chromosome names without the prefix \code{"chr"} which is used by UCSC).

<<SequenceTrackFromFaFile, results=hide, eval=FALSE>>=
## It is important to disable the UCSC chromosome naming!
options(ucscChromosomeNames=FALSE)

## Create the SequenceTrack.
faTrack <- SequenceTrack(sequence=dna, chromosome="2", genome="GRCh38")

## Plot it.
plotTracks(list(GenomeAxisTrack(), faSeq), from=199970, to=200000)
@


\subsubsection*{Display parameters for SequenceTrack objects}

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{SequenceTrack}
class by typing in \code{?SequenceTrack} on the \RR command line.

<<SequenceTrackClassTableSec, echo=FALSE, results=tex>>=
addParTable("SequenceTrack")
@

\subsection{AlignmentsTrack}
Plots of aligned sequences, typically from next generation sequencing
experiments can be quite helpful, for instance when visually
inspecting the validity of a called SNP. Those alignments are usually
stored in \code{BAM} files.  As shown in some of the previous section
we can deal with these files in a fairly crude way by coupling them to
either an \Rclass{AnnotationTrack} or a \Rclass{DataTrack}. The utility
of the generated plots however is fairly limited, and we want in
particular support for gapped and paired aligments, and show
differences of the individual read sequences to a reference.

The \Rclass{AlignmentsTrack} class is designed to help in exactly
these tasks. In its typical manifestation it is also a member of the
\Rclass{ReferenceTrack} class, enabling memory efficient streaming
directly off a file. Even though this file does not necessarily have
to be a \code{BAM} file, the automatic import will only work for
those. For all other file representations of alignments a user will
have to provide their own import function, and this topic is further
discussed in the class' documentation. For this demonstation let's use
a small \code{BAM} file for which paired NGS reads have been mapped to
an extract of the human hg19 genome. The data originate from an RNASeq
experiment, and the alignements have been performed using the
\code{STAR} aligner allowing for gaps. We also download some gene
annotation data for that region from Biomart.

Again, creating the \Rclass{AlignmentsTrack} object is as simple as
calling the constructor function of the same name. We set the
\Rfunarg{isPaired} argument to \code{TRUE} in order to tell the
constructor that it should expect paired reads. Please note that
\code{TRUE} is the default value for this parameter and usually things
will still work even if the reads do not originate from a paired end
library. In other words, the default setting could be used as a
poor-man's autodetection, however it make more sense to supply the
argument explicitely if you know what type of data to expect.

<<alignmentstrack_1_do, echo=FALSE, results=hide>>=
afrom <- 2960000
ato <- 3160000
alTrack <- AlignmentsTrack(system.file(package="Gviz", "extdata", "gapped.bam"), isPaired=TRUE)
data(alTrackGenes)
@

<<alignmentstrack_1_show, eval=FALSE>>=
afrom <- 2960000
ato <- 3160000
alTrack <- AlignmentsTrack(system.file(package="Gviz", "extdata", "gapped.bam"), isPaired=TRUE)
bmt <- BiomartGeneRegionTrack(genome="hg19", chromosome="chr12", start=afrom, end=ato,
                              filter=list(with_ox_refseq_mrna=TRUE), stacking="dense")
@

We can create a first plot of our read alignments data from a birds
eye perspecive, say a range of 20,000 nucleotides.

<<alignmentstrack_2, results=hide, width=7.5, height=5, fig=TRUE>>=
plotTracks(c(bmt, alTrack), from=afrom, to=ato, chromosome="chr12")
@

Now this already shows us the general layout of the track: on top we
have a panel with the read coverage information in the form of a
histogram, and below that a pile-up view of the individual
reads. There is only a certain amount of vertical space available for
the plotting, and not the whole depth of the pile-up can be displayed
here. This fact is indicated by the white downward-pointing arrows in
the title panel. We could address this issue by playing around with
the \Rfunarg{max.height}, \Rfunarg{min.height} or
\Rfunarg{stackHeight} display parameters which all control the height
or the vertical spacing of the stacked reads. Or we could reduce the
size of the coverage section by setting the \Rfunarg{coverageHeight}
or the \Rfunarg{minCoverageHeight} parameters. Please see the class
documentation for more details.

<<alignmentstrack_3, results=hide, width=7.5, height=5, fig=TRUE>>=
plotTracks(c(bmt, alTrack), from=afrom, to=ato, chromosome="chr12", min.height=0, coverageHeight=0.08,
           minCoverageHeight=0)
@

From that far out the pile-ups are not particlularly useful, and we
can turn those off by setting the \Rfunarg{type} display parameter
accordingly.

<<alignmentstrack_4, results=hide, width=7.5, height=2, fig=TRUE>>=
plotTracks(c(alTrack, bmt), from=afrom, to=ato, chromosome="chr12", type="coverage")
@

Let's zoom in a bit further to check out the details of the pile-ups
section.

<<alignmentstrack_5, results=hide, width=7.5, height=5, fig=TRUE>>=
plotTracks(c(bmt, alTrack), from=afrom+12700, to=afrom+15200, chromosome="chr12")
@

The direction of the individual  reads is indicated by the arrow head,
and  read  pairs are  connected by  a bright  gray  line.  Gaps in
the alignments are show by the connecting dark gray lines. Numbers of
reads supporting each gap can be visualized as ``sashimi plot''
\cite{Katz22012015} by adding \Rfunarg{``sashimi''} as an additional
agument to the \Rfunarg{type} display parameter.

<<alignmentstrack_5_1, results=hide, width=7.5, height=2, fig=TRUE>>=
plotTracks(c(bmt, alTrack), from=afrom+12700, to=afrom+15200, chromosome="chr12",
           type=c("coverage", "sashimi"))
@

It is also possible to filter the junctions by including the
coordinates as an additional argument \Rfunarg{``sashimiFilter''}. In
this case only junction which equally overlaps with specified
\Rclass{GRanges} object are shown.

<<alignmentstrack_5_2, results=hide, width=7.5, height=2, fig=TRUE>>=
introns <- GRanges("chr12", IRanges(start=c(2973662, 2973919), end=c(2973848, 2974520)))
plotTracks(c(bmt, alTrack), from=afrom+12700, to=afrom+15200, chromosome="chr12",
           type=c("coverage", "sashimi"), sashimiFilter=introns)
@

However sometimes the aligners do not place the junction reads
precisely on the exon-intron boundaries and it is therefore useful to
include also junctions which are shifted by few bases. This can be done
by setting the additional argument
\Rfunarg{``sashimiFilterTolerance''} to small positive integer value.

<<alignmentstrack_5_3, results=hide, width=7.5, height=2, fig=TRUE>>=
plotTracks(c(bmt, alTrack), from=afrom+12700, to=afrom+15200, chromosome="chr12",
           type=c("coverage", "sashimi"), sashimiFilter=introns, sashimiFilterTolerance=5L)
@

On devices that support transparancy we  can also see that some of
the read pairs are actually overlapping. Like  before, we can not
display  the full depth of the pile-up,  and the clipping that takes
place hides  a lot of the reads from  the view. Another way of
getting to see at  least some of them is by reversing the stacking
order. We can also play around with some of the other display
parameters that control the look and feel of the reads.

<<alignmentstrack_6, results=hide, width=7.5, height=5, fig=TRUE>>=
plotTracks(c(bmt, alTrack), from=afrom+12700, to=afrom+15200, chromosome="chr12", reverseStacking=TRUE,
           col.mates="purple", col.gap="orange", type="pileup")
@

As mentioned before we can control whether the data should be treated
as paired end or single end data by setting the \Rfunarg{isPaired}
argument in the constructor. Here is how we could take a look at the
data in the same file, but in single end mode.

<<alignmentstrack_6_1, results=hide, width=7.5, height=5, fig=TRUE>>=
alTrack <- AlignmentsTrack(system.file(package="Gviz", "extdata", "gapped.bam"), isPaired=FALSE)
plotTracks(c(bmt, alTrack), from=afrom+12700, to=afrom+15200, chromosome="chr12")
@

So far we did not look at the individual read sequences at all. Even
though they are contained in the \code{BAM} file, their value is
fairly limited without the reference sequence against which the reads
have initially been aligned. Typically the mismatch positions and the
variant bases are not included. (Please note that this is not entirely
true for all \code{BAM} files. One can store base mismatch positions
in the CIGAR string, or even the full variant description in the
non-standard \code{MD} field. However very few aligners currently make
use of these features. We may implement support for those later if
found necessary.)

To better show the features of the \Rclass{AlignmentsTrack} for
sequence variants we will load a different data set, this time from a
whole genome DNASeq SNP calling experiment. Again the reference genome
is hg19 and the alignments have been performed using \code{Bowtie2}.

<<alignmentstrack_7, results=hide, width=7.5, height=5, fig=TRUE>>=
afrom <- 44945200
ato <- 44947200
alTrack <- AlignmentsTrack(system.file(package="Gviz", "extdata", "snps.bam"), isPaired=TRUE)
plotTracks(alTrack, chromosome="chr21", from=afrom, to=ato)
@

Somehow we need to tell the \Rclass{AlignmentsTrack} about the
reference genome, and there are two ways to archive this. Either we
provide a \Rclass{SequenceTrack} object with the reference to the
constructor function via the \Rfunarg{referenceSequence} argument, or
we stick it into the track list that is provided to the
\Rfunction{plotTrack} which is smart enough to detect it there. For
this example we chose the second option and re-use the
\Rclass{SequenceTrack} object from the previous section.

<<alignmentstrack_8, results=hide, width=7.5, height=5, fig=TRUE>>=
plotTracks(c(alTrack, sTrack), chromosome="chr21", from=afrom, to=ato)
@

The mismatched bases are now indicated on both the individual reads in
the pileup section and also in the coverage plot in the form of a
stacked histogram. When zooming in to one of the obvious heterozygous
SNP positions we can reveal even more details.

<<alignmentstrack_9, results=hide, width=7.5, height=5, fig=TRUE>>=
plotTracks(c(alTrack, sTrack), chromosome="chr21", from=44946590, to=44946660)
@

Just like shown before in the \Rclass{SequenceTrack} we can even
show individual letters here as long as there is enough space to fit
them. This can be controlled using the display parameters
\code{min.height}, \code{max.height} and \code{cex}.

<<alignmentstrack_10, results=hide, width=7.5, height=5, fig=TRUE>>=
plotTracks(c(alTrack, sTrack), chromosome="chr21", from=44946590, to=44946660, cex=0.5, min.height=8)
@

\subsubsection*{Display parameters for AlignmentsTrack objects}

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{AlignmentsTrack}
class by typing in \code{?AlignmentsTrack} on the \RR command line.

<<AlignmentsTrackClassTable, echo=FALSE, results=tex>>=
addParTable("AlignmentsTrack")
@


\subsection{Creating tracks from UCSC data}

The UCSC data bases contain a multitude of genome annotation data for
dozents of different organisms. Some of those data are very simple
annotations like CpG island locations or SNP locations. Others are
more complicated gene models, or even numeric annotations like
conservation information. In order to provide a unified interface to
all this information, the \mgg package defines a meta-constructor
function \Rfunction{UcscTrack}. The idea here is that we can express
all of the available Ucsc data in one of the package's track types. We
use the functionality provided in the \Rpackage{rtracklayer} package
to connect to UCSC and to download the relevant information. As a
little illustrative example, let's reproduce a view from the famous
UCSC genome browser using the \mgg package. As a final result we want
to show something similar to Figure~\ref{fig:UCSC1}.

\begin{figure}[htb]
  \centering
  \includegraphics{ucsc1.pdf}
  \label{fig:UCSC1}
  \caption{A screen shot of a UCSC genome browser view around the FMR1
    locus on the mouse chromosome.}
\end{figure}

To start we first need to know about the available data in the UCSC
data base and about their structure. A good way to do this is to use
the table browser on the UCSC web site
(\url{http://genome.ucsc.edu/cgi-bin/hgTables?command=start}). Figure~\ref{fig:UCSC2}
shows the table structure for the first gene model track, the known
UCSC genes, in the table browser. We can see that there are multiple
fields, some with genomic locations, other with additional data like
labels or identifiers. If we go back to the section about the
\Rclass{GeneRegionTrack} class we remember that we need exactly this
type of information for the constructor function. So in order to take
the UCSC data and build an object of class \Rclass{GeneRegionTrack} we
need a way to map them to the individual constructor arguments. This
is exactly what the \Rfunction{UcscTrack} meta-constructor is supposed
to do for us.

\begin{figure}[htb]
  \centering
  \includegraphics{ucsc2.pdf}
  \label{fig:UCSC2}
  \caption{A screen shot of a UCSC table browser view on the UCSC Known Genes track.}
\end{figure}

It needs to know about the track for which to extract the data (and
optionally one or several of the tables that make up the collective
track data, see \code{?UcscTrack} for details), about the genomic
range including the chromosome for which to extract data, about the
type of \mgg track that we want to translate this data into, and about
the individual track columns and their counterparts in the respective
track class constructor. In our example, the track is called
\code{knownGene}, the track type to construct is
\code{GeneRegionTrack}, and the relevant columns are
\code{exonStarts}, \code{exonEnds}, \code{name} and \code{strand},
which we will use as the start and end coordinates of the ranges and
for all the exon, transcript and gene identifiers. Here we make use of
the high flexibility of the \Rfunction{GeneRegionTrack} constructor in
the sense that the exon coordinates actually come in the form of a
comma-separated list, combining all the information for one transcript
in one row of the table. The function is smart enough to detect this
and to split the annotation regions accordingly. The full function
call to create the \Rclass{GeneRegionTrack} from the UCSC data looks
like this:

<<ucscTrack1, eval=FALSE>>=
from <- 65921878
to <- 65980988
knownGenes <- UcscTrack(genome="mm9", chromosome="chrX", track="knownGene", from=from, to=to,
                        trackType="GeneRegionTrack", rstarts="exonStarts", rends="exonEnds", gene="name",
                        symbol="name", transcript="name", strand="strand", fill="#8282d2", name="UCSC Genes")
@

With a similar approach we can construct the next two gene model
tracks based on the \code{xenoRefGene} and \code{ensGene} data tables.

<<ucscTrack2, eval=FALSE>>=
refGenes <- UcscTrack(genome="mm9", chromosome="chrX", track="xenoRefGene", from=from, to=to,
                      trackType="GeneRegionTrack", rstarts="exonStarts", rends="exonEnds", gene="name",
                      symbol="name2", transcript="name", strand="strand", fill="#8282d2",
                      stacking="dense", name="Other RefSeq")

ensGenes <- UcscTrack(genome="mm9", chromosome="chrX", track="ensGene", from=from, to=to,
                      trackType="GeneRegionTrack", rstarts="exonStarts", rends="exonEnds", gene="name",
                      symbol="name2", transcript="name", strand="strand", fill="#960000",
                      name="Ensembl Genes")
@

The CpG and SNP tracks are slightly different since a
\Rclass{GeneRegionTrack} representation would not be particularly
useful. Instead, we can use \Rclass{AnnotationTrack} objects as
containers. The overall process using the \Rfunction{UcscTrack}
meta-constructor remains the same.

<<ucscTrack3, eval=FALSE>>=
cpgIslands <- UcscTrack(genome="mm9", chromosome="chrX", track="cpgIslandExt", from=from, to=to,
                        trackType="AnnotationTrack", start="chromStart", end="chromEnd", id="name",
                        shape="box", fill="#006400", name="CpG Islands")

snpLocations <-  UcscTrack(genome="mm9", chromosome="chrX", track="snp128", from=from, to=to,
                           trackType="AnnotationTrack", start="chromStart", end="chromEnd", id="name",
                           feature="func", strand="strand", shape="box", stacking="dense", fill="black",
                           name="SNPs")
@

Most of UCSC's \Rclass{DataTrack}-like tracks are a little more
complex and represent a collection of several sub-tracks, with data
originating from multiple tables. To make sure that we get the correct
information we have to be a little bit more specific here and also
define the particular table on the UCSC data base to use.

<<ucscTrack4, eval=FALSE>>=
conservation <- UcscTrack(genome="mm9", chromosome="chrX", track="Conservation", table="phyloP30wayPlacental",
                          from=from, to=to, trackType="DataTrack", start="start", end="end", data="score",
                          type="hist", window="auto", col.histogram="darkblue", fill.histogram="darkblue",
                          ylim=c(-3.7, 4), name="Conservation")

gcContent <- UcscTrack(genome="mm9", chromosome="chrX", track="GC Percent", table="gc5Base",
                       from=from, to=to, trackType="DataTrack", start="start", end="end", data="score",
                       type="hist", window=-1, windowSize=1500, fill.histogram="black", col.histogram="black",
                       ylim=c(30, 70), name="GC Percent")
@

To add some reference points we also need a genome axis and an
\Rclass{IdeogramTrack} of the x chromosome.

<<ucscTrack5, eval=FALSE>>=
axTrack <- GenomeAxisTrack()
idxTrack <- IdeogramTrack(genome="mm9", chromosome="chrX")
@

And finally we can plot all of our tracks.

<<ucscTrackLoad, echo=FALSE, results=hide>>=
data(ucscItems)
@

<<ucscTrack6, fig=TRUE, results=hide, width=7.5, height=5.5>>=
plotTracks(list(idxTrack, axTrack, knownGenes, refGenes, ensGenes, cpgIslands,
                gcContent, conservation, snpLocations), from=from, to=to, showTitle=FALSE)
@

\section{Track highlighting and overlays}

\subsection{Highlighting}

Some features of a typical genome browser plot are not specific for
just a single track, but instead span multiple tracks. One prominent
example are highlighting regions. Assume you want to draw the
attention on the plot to a particular genomic stretch where the visual
clues across serveral or maybe even all tracks indicate a particular
interesting behaviour. Because the individual tracks in the \mgg
package are all represented by individual objects, there is no
straight forward way to define such a region. Here is where the
\Rclass{HighlightingTrack} class comes into play. Objects from this
class are essentially just containers for one or several of the
previously described track objects, togehter with a very simple range
definition that is to be superimposed over the individual tracks once
they have been rendered. Going back to our final example from the
second chapter we could highlight two areas on the plot on the
\Rclass{AnnotationTrack}, the \Rclass{GeneRegionTrack} and the
\Rclass{DataTrack} using the \Rfunction{HighlightTrack} constructor in
the following call:

<<HighlightTrack, fig=TRUE, width=7.5, height=4>>=
ht <- HighlightTrack(trackList=list(atrack, grtrack, dtrack), start=c(26705000, 26720000), width=7000, chromosome=7)
plotTracks(list(itrack, gtrack, ht), from = lim[1], to = lim[2])
@

One can also build multiple \Rclass{HighlightTrack} collections to
create gaps in the highlighting regions. For \Rclass{IdeogramTracks}
we never plot highlighting ranges because they do not live on the same
x-axis coordinate system like the other tracks.

<<HighlightTrack2, fig=TRUE, width=7.5, height=4>>=
ht1 <- HighlightTrack(trackList=list(itrack, gtrack, atrack), start=c(26705000, 26720000), width=7000, chromosome=7)
ht2 <- HighlightTrack(trackList=dtrack, start=c(26705000, 26720000), width=7000, chromosome=7)
plotTracks(list(ht1, grtrack, ht2), from=lim[1], to=lim[2])
@

\subsubsection*{Display parameters for HighlightTrack objects}

For a complete listing of all the available display parameters please
see the table below or the man page of the \Rclass{HighlightTrack}
class by typing in \code{?HighlightTrack} on the \RR command line.

<<HighlightTrackClassTable, echo=FALSE, results=tex>>=
addParTable("HighlightTrack")
@

\subsection{Overlays}

For certain applications it can make sense to overlay multiple tracks
on the same area of the plot. For instance when dealing with multiple
data sets that do not necessarily share the same genomic coordinates
for their data points the design of the \Rclass{DataTrack} class is
not optimal. To contrast these two data sets an overlay of two
separate \Rclass{DataTrack} objects however can be quite powerful.The
\mgg package allows for such overlay operations by means of the
\Rclass{OverlayTrack} class.Similar to the \Rclass{HighlightTrack}
class from the previous section this is merely a container for an
arbitrary number of other \mgg track objects. For the purpose of an
instructive example we will generate a second \Rclass{DataTrack}
object and combine it with the existing one from the second chapter.

<<OverlayTrack, fig=TRUE, width=7.5, height=3>>=
dat <- runif(100, min=-2, max=22)
dtrack2 <- DataTrack(data=dat, start=coords[-length(coords)], end=coords[-1], chromosome=chr,
                     genome=gen, name="Uniform2", groups=factor("sample 2", levels=c("sample 1", "sample 2")),
                     legend=TRUE)
displayPars(dtrack) <- list(groups=factor("sample 1", levels=c("sample 1", "sample 2")), legend=TRUE)
ot <- OverlayTrack(trackList=list(dtrack2, dtrack))
ylims <- extendrange(range(c(values(dtrack), values(dtrack2))))
plotTracks(list(itrack, gtrack, ot), from=lim[1], to=lim[2], ylim=ylims, type=c("smooth", "p"))
@

It is important to know that \mgg is fully unaware of the intention
and any specifics of the track merging operation. Any track type can
be merged with any other track type, and there are no limitations on
the number of overlays. It is totally up to the user to design
reasonable merges and to control critical aspects of the plot, like
the y-axis scaling or the grouping in the above example. During
rendering, the software will only use the first track object in the
\Robject{trackList} container of \Rclass{OverlayTracks} object to set
up the track title. Thus, the ranges on the y-axis are those that have
been automatically determined for the first of our two
\Rclass{DataTracks}. Only by forcing a common y-range for both objects
do we actually plot the data from both tracks on a common scale. Had
we added another track object that is not a \Rclass{DataTrack} is the
first element of our \Rclass{trackList} there wouldn't even be an
axis. In a similar fashion, by nicely setting up the group levels for
the two \Rclass{DataTracks} we can even ask for a legend to be added.

On devices that support it, alpha blending can be a useful tool to
tease out even more information out of track overlays, at least when
comparing just a small number of samples. The resulting transparancy
effctively eliminates the probelm of overplotting. The following
example will only work if this vignette has been built on a system
with alpha blending support.

<<OverlayTrack2, fig=TRUE, width=7.5, height=3>>=
displayPars(dtrack) <- list(alpha.title=1, alpha=0.5)
displayPars(dtrack2) <- list(alpha.title=1, alpha=0.5)
ot <- OverlayTrack(trackList=list(dtrack, dtrack2))
plotTracks(list(itrack, gtrack, ot), from=lim[1], to=lim[2], ylim=ylims, type=c("hist"), window=30)
@



\section{Composite plots for multiple chromosomes}

As mentioned in the introduction section, a set of \mgg tracks has to
share the same chromosome when plotted, i.e., only a single chromosome
can be active during a given plotting operation. Consequently, we can
not directly create plots for multiple chromosomes in a single call to
the \Rfunction{plotTracks} function. However, since the underlying
graphical infrastructure of the \mgg package uses grid graphics, we
can build our own composite plot using multiple consecutive
\Rfunction{plotTracks} calls. All we need to take care of is an
adequate layout structure to plot into, and we also need to tell
\Rfunction{plotTracks} not to clear the graphics device before
plotting, which can be archieved by setting the function's
\Rfunarg{add} argument to \code{FALSE}. For details on how to create a
layout structure in the grid graphics system, please see the help page
at \code{?grid.layout)}.

We start by creating an \Rclass{AnnotationTrack} objects and a
\Rclass{DataTrack} object which both contain data for several
chromosomes.

<<multPlot1>>=
chroms <- c("chr1", "chr2", "chr3", "chr4")
maTrack <- AnnotationTrack(range=GRanges(seqnames=chroms, ranges=IRanges(start=1, width=c(100,400,200,1000)),
                                         strand=c("+", "+", "-", "+")), genome="mm9", chromosome="chr1", name="foo")

mdTrack <- DataTrack(range=GRanges(seqnames=rep(chroms, c(10, 40, 20, 100)),
                                   ranges=IRanges(start=c(seq(1,100,len=10), seq(1,400,len=40), seq(1, 200, len=20),
                                                          seq(1,1000, len=100)), width=9), values=runif(170)),
                     data="values", chromosome="chr1", genome="mm9", name="bar")
@

Now we also want a genome axis and an \Rclass{IdeogramTrack} object to indicate the genomic context.

<<multPlot2>>=
mgTrack <- GenomeAxisTrack(scale=50, labelPos="below", exponent=3)
chromosome(itrack) <- "chr1"
@

Finaly, we build a layout in which the plots for each chromosome are
placed in a rectangular grid and repeatedly call
\Rfunction{plotTracks} for each chromosome.

<<multPlot3, fig=TRUE, results=hide, width=7.5, height=4>>=
ncols <- 2
nrows <- length(chroms)%/%ncols
grid.newpage()
pushViewport(viewport(layout=grid.layout(nrows, ncols)))
for(i in seq_along(chroms)){
    pushViewport(viewport(layout.pos.col=((i-1)%%ncols)+1, layout.pos.row=(((i)-1)%/%ncols)+1))
    plotTracks(list(itrack, maTrack, mdTrack, mgTrack), chromosome=chroms[i], add=TRUE)
    popViewport(1)
}
@

Maybe an even more compact version of this would be to use the lattice
package for building the actual trellis, with \Rfunction{plotTracks}
as the panel function.

<<multPlot4, fig=TRUE, results=hide, width=7.5, height=4>>=
library(lattice)
chroms <- data.frame(chromosome=chroms)
xyplot(1~chromosome|chromosome, data=chroms, panel=function(x){plotTracks(list(itrack , maTrack, mdTrack, mgTrack), chromosome=x, add=TRUE, showId=FALSE)},
       scales=list(draw=FALSE), xlab=NULL, ylab=NULL)
@


\section{Bioconductor integration and file support}
This short section is supposed to give a very brief overview over the
different track classes in the \mgg package and how those can be
constructed from the typical Bioconductor classes that deal with
genomic data. The list ist by no means complete, and a closer look at
a track class' documentation should provide all the possible options.

\begin{longtable}{ l | l | p{9.5cm} }
  \hline
  Gviz class & Bioconductor class & Method\\
  \hline
  AnnotationTrack & data.frame & Constructor \\
   & IRanges & Constructor + additional arguments \\
   & GRanges & Constructor or setAs method, additional data in metadata columns \\
   & GRangesList & Constructor or setAs method \\
  \hline
  GeneRegionTrack & data.frame & Constructor \\
   & IRanges & Constructor + additonal arguments \\
   & GRanges & Constructor or setAs method, additional data in metadata columns \\
   & GRangesList & Constructor or setAs method, additional data in metadata columns \\
   & TxDb & Constructor or setAs method \\
   & EnsDb & Constructor \\
  \hline
  DataTrack & data.frame & Constructor \\
   & IRanges & Constructor + additional data matrix \\
   & GRanges & Constructor or setAs method, numeric data in metadata columns \\
  \hline
  SequenceTrack & DNAStringSet & Constructor \\
   & BSgenome & Constructor \\
   & FaFile & Constructor \\
  \hline
\end{longtable}

We have seen in previous examples that the \mgg package supports quite
a wide range of standard file types as inputs for the class'
constructor functions. In this section we summarize the file support
for the different track classes in one place.

\begin{longtable}{ l | l | l | c | p{6.5cm} }
  \hline
  Gviz class & File type & Extension & Streaming & Details\\
  \hline
  AnnotationTrack & BED & \code{.bed} & -- & Genomic locations from the mandatory \code{chrom}, \code{chromStart} and \code{chromEnd} fields, and optionally the strand from the \code{strand} field. If present, the information in the \code{name} field is mapped to track item ids, and \code{itemRgb} is mapped to track item feature type. All other fields are currently ignored.\\
  & GFF & \code{.gff}, \code{.gff1} & -- & Only the following basic GFF fields are recognized: \code{seqname}, \code{start}, \code{end}, \code{end}, \code{strand}, \code{feature} (mapped to track item feature type) and \code{group} (to allow for track item grouping). \\
  & GFF2 & \code{.gff2} & -- & Same as above, but feature grouping information may be provided either as \code{Group} or \code{Parent} attribute. Feature ids are mapped to one of the \code{ID}, \code{Name} or \code{Alias} attributes.\\
  & GFF3 & \code{.gff3} & -- & Same as above, but feature grouping information has to be provided as the \code{Parent} attribute.\\
  & BAM & \code{.bam} & $\surd$ & Only start and end locations as well as the strand information for the reads are used. Read identifiers are used for track item grouping.\\
  \hline
  GeneRegionTrack & GTF & \code{.gtf} & -- & A somewhat looser format definition for \code{gtf} files is applied here where gene, transcript and exon identifiers and names can be parsed from the \code{gene\_id}, \code{gene\_name}, \code{transcript\_id}, \code{transcript\_name}, \code{exon\_id} or \code{exon\_id} attributes.\\
  & GFF & \code{.gff}, \code{.gff1} & -- & This only supports very limited item grouping and thus complete gene models can not be properly encoded.\\
  & GFF2 & \code{.gff2} & -- & In most instances this is identical to the \code{GTF} standard and it could make sense to rename the file accordingly.\\
  & GFF3 & \code{.gff3} & -- & The gene-to-transcript and transcript-to-exon relationships are encoded in the \code{parent} and \code{type} attributes and the parser tries to accomodate most of the exisiting GFF3 variants.\\
  \hline
  DataTrack & BedGraph & \code{.bedGraph} & -- & \\
  & WIG & \code{.wig} & -- & \\
  & BigWig & \code{.bigWig}, \code{.bw} & $\surd$ & \\
  & BAM & \code{.bam} & $\surd$ & Read coverage only is extracted from the \code{bam} file.\\
  \hline
  SequenceTrack & FASTA & \code{.fa}, \code{fasta} & $(\surd)$ & Streaming only possible if an index file is found in the same directory as the original fasta file.\\
  & 2Bit & \code{.2bit} & $\surd$ & \\
  \hline
  AlignmentsTrack & BAM & \code{.bam} & $\surd$ & ALways needs an index file is found in the same directory as the original BAM file.\\
  \hline
\end{longtable}



\clearpage
\section*{SessionInfo}
The following is the session info that generated this vignette:
<<session-info>>=
  sessionInfo()
@

\clearpage
\bibliography{Gviz-refs}

\end{document}
